{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eabedb5d-7fb2-4665-8b77-330ac9993e56",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h1>Alzheimer's, Genetics, and the Microbioime-Gut-Brain Axis:<br><br> How Naturopathic Medicine Affects Alzheimer's (Part One: Data Wrangling)</h1>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57be9d6e-01d9-42d1-b26d-f8c722a4d042",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h2>1.1 Introduction</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828797b9-f5eb-4966-a95f-4bbbd2fd19b2",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <h3>1.1.1 Context of Capstone</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "*The United States Secretary of the Department of Health has declared for many years that the public health crisis of Alzheimer’s is an ongoing federal problem that US citizens face. The gravity of this disease is very serious and can be fatal. Today, “nearly 7 million Americans live with Alzheimer’s” <a href=\"#mla_1\">[1]</a>\n",
    ", and “the cost of care yearly is 360 billion USD” <a href=\"#mla_2\">[2]</a>. “10% of adults aged 45 and older report subject cognitive decline” <a href=\"#mla_3\">[3]</a>, which is the earliest warning sign of Alzheimer’s disease. “By 2050, nearly 13 million Americans could be living with Alzheimer’s <a href=\"#mla_4\">[4]</a>, where costs will reach nearly $1 trillion” <a href=\"#mla_5\">[5]</a>*.\n",
    "\n",
    "Figure. 1 below is taken from the \"2024 Alzheimer’s Disease\n",
    "Facts and Figures\" report from the Alzheimer's Association® <a href=\"#mla_6\">[6]</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef16dc4-7c83-44d9-a81f-001f2eda1873",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"AlzProjection.png\" alt=\"Fig. 1\" width=\"90%\">\n",
    "    <p style=\"text-align: center;\">Figure. 1 shows the Alzheimer's Disease projection by year 2060.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "558aef5f-5104-45b3-bff0-5b57c69e9a95",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>What is Naturopathic Medicine?</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Conventional medicine is a system designed for a brief consultation with a medical doctor, typically lasting 15 minutes. The goal is to diagnose as quickly as possible and to treat ailments or diseases with prescribed pharmaceuticals. However, this approach often fails to address the root causes of systemic issues affecting a patient's whole body.\n",
    "\n",
    "Conventional medicine struggles to address the public health crises facing Americans. While it is \"highly effective in addressing acute and life-threatening conditions\" <a href=\"#mla_7\">[7]</a> through surgery and necessary medications, it often falls short as a definitive model for combating public health crises such as Alzheimer’s. This limitation contributes to problems such as the opioid crisis and increased rates of ADHD. Although pharmaceuticals are essential in many cases, they are not a comprehensive solution for addressing the functional and holistic health of a patient's entire body, which influences a patient's disease status.\n",
    "\n",
    "This is where Naturopathic Medicine comes in. This medicine is a science-based paradigm that aims to treat the whole body by identifying and addressing the root causes of disease. It seeks to improve all health markers of a patient's bodily systems through protocols involving nutraceuticals, diet, nutrition, health tests, lifestyle therapies, and natural treatments, allowing a patient's body to heal itself completely or to its fullest extent. Naturopathic Medicine \"was founded in the United States in 1901 by Benedict Lust\"<a href=\"#mla_8\"> [8]</a>, taken from a movement that was happening at the turn of the century in Europe.  \n",
    "\n",
    "Figure. 2 below depicts the fundamental principles of Naturopathic Medicine. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1dedae3-5fb1-481e-a4f1-8d8128e037d2",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"Principals%20of%20NM.png\" alt=\"Fig. 1\" width=\"40%\">\n",
    "    <p style=\"text-align: center;\">Figure. 2 shows the 6 Principles of Naturopathic Medicine<a href=\"#mla_105\"> [105]</a>.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15ea933f-d28b-49e8-8609-9acab3a95cf0",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>Therapeutic Order and Course of Action in Naturopathic Medicine</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "<ol>\n",
    "<li><strong>First Layer:</strong> <i>Establish the Foundation for Optimal Health:</i> A Naturopathic Doctor (ND) educates patients on the basics of living healthy, laying the groundwork for long-term health.<br><br></li>\n",
    "\n",
    "<li><strong>Second Layer:</strong> <i>Stimulate the Body's Natural Healing Power:</i> Initially, an ND prescribes supplements or nutraceuticals to address inflammatory or oxidative \"bottlenecks.\" This helps alleviate high levels of inflammation contributing to an ailment<a href=\"#mla_9\"> [9]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Third Layer:</strong> <i>Support & Restore the Body's Weakened Systems:</i> After the baseline education, supplements, and nutraceuticals to handle the major sources of inflammation that contribute to a patient's ailments, protocols and programs of supplements, nutraceuticals, natural therapies, diet, nutrition, and lifestyle changes seek to naturally restore a body's weakened systems completely, or to their fullest possible extent<a href=\"#mla_10\"> [10]</a>. The following are some biomarker tests that assess the health of bodily systems, along with an example of how they may be weakened. They are labeled from first to last in layers like the pyramid in Figure. 3 as shown below, to also indicate what is more fundamental to overall health.</i><br><br>\n",
    "</ol>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <h3>Bodily Health Systems: Their Sequential Importance and Interconnection</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "<ol>\n",
    "<li><strong>First Layer (Bottom of the Pyramid):</strong> <i>Gastrointestinal Tract:</i> A stool test assesses the gut microbiome and intestinal health markers to identify toxic burdens and inflammation from bacteria, pathogens, infections, and overall digestion<a href=\"#mla_11\"> [11]</a><a href=\"#mla_12\"> [12]</a><a href=\"#mla_13\"> [13]</a></a>.<br><br></li>\n",
    "\n",
    "<li><strong>Second Layer:</strong> <i>Liver and Detox Tract:</i> A hair follicle and mold test assesses heavy metal toxicity and mold levels in a person to evaluate how well a patient is managing oxidative stress and detoxification, which contribute to aging and ailments<a href=\"#mla_14\"> [14]</a><a href=\"#mla_15\"> [15]</a><a href=\"#mla_16\"> [16]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Third Layer:</strong> <i>Hormonal System:</i> A saliva test that evaluates reproductive hormones' ratios and quantities, which are crucial for inflammation and overall wellbeing. Example: A woman who has chronically low levels of estrogen, which affects her mood and inflammation levels<a href=\"#mla_17\"> [17]</a><a href=\"#mla_18\"> [18]</a><a href=\"#mla_19\"> [19]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Fourth Layer:</strong> <i>Adrenal System:</i> A day-long saliva test measures cortisol levels throughout the day to assess adrenal fatigue, better known as \"burnout\"<a href=\"#mla_20\"> [20]</a><a href=\"#mla_21\"> [21]</a><a href=\"#mla_22\"> [22]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Fifth Layer:</strong> <i>Neurotransmitter Test:</i> A saliva panel evaluates the ratios of neurotransmitters, essential for mental wellbeing<a href=\"#mla_23\"> [23]</a><a href=\"#mla_24\"> [24]</a><a href=\"#mla_25\"> [25]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Sixth Layer (Top of the Pyramid):</strong> <i>Genetics Test:</i> A saliva test assesses which genes may contain mutations, potentially indicating \"dirty genes\" needing to be cleaned up. Improvement of the first layer health system improves markers in all subsequent layers. Improving all health system layers before the Genetics Layer consequentially improves the status of the genetics layer<a href=\"#mla_26\"> [26]</a><a href=\"#mla_27\"> [27]</a><a href=\"#mla_28\"> [28]</a>.<br><br></li>\n",
    "</ol>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "<ol start=\"4\">\n",
    "    <li><strong>Fourth Layer:</strong> <i>Targeted Natural Prescriptions:</i> After weakened bodily systems are completely restored or restored to their fullest possible extent, if there is still a deficiency within the health systems of a patient or other kinds of deficiencies and bottlenecks, then whatever the nutraceuticals from the second layer cannot address is addressed by the addition of targeted natural prescriptions<a href=\"#mla_29\"> [29]</a>.<br><br></li>\n",
    "\n",
    "<li><strong>Fifth Layer:</strong> <i>Replacement Therapy:</i> After targeted natural prescriptions are issued, there might still be deficiencies within a patient. For example, testosterone-boosting supplements such as Fadogia Agrestis and Tongkat Ali are potent natural interventions, but a male patient still suffering from low testosterone will benefit from testosterone replacement therapy<a href=\"#mla_30\"> [30]</a><a href=\"#mla_31\"> [31]</a>. This is a recommended course of action to address this highly inflammatory imbalance.<br><br></li>\n",
    "\n",
    "<li><strong>Sixth Layer:</strong> <i>Drug Therapy:</i> At this stage, if there is still a need to address imbalances within a patient, prescription pharmaceuticals are recommended.<br><br></li>\n",
    "\n",
    "<li><strong>Final Layer:</strong> <i>Surgery:</i> The final stage is invasive surgery to address any structural or biological imbalances that require this level of attention.<br><br></li>\n",
    "</ol>\n",
    "\n",
    "<strong><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">Important Note: Since we are a whole body and all of our bodily systems are interconnected as a whole, a heavier inflammatory load in the first layer is a heavier load on all of the following sequential layers. A heavier inflammatory and oxidative load on the fourth layer is a heavier load for the fifth through last layers, etc.</blockquote><br>\n",
    "\n",
    "<strong><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">\"Conventional medicine typically only addresses replacement therapy, drug therapy, and surgery while ignoring the first four layers of fundamental health.\"</blockquote><be>\n",
    "\n",
    "*Figure. 3 below depicts the Therapeutic order of Naturopathic Medicine.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62b77b64-4a22-4ebc-b86b-077b3f69a3d8",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"NM%20Therapeutic%20Order.png\" alt=\"Fig. 2\" width=\"55%\">\n",
    "    <p style=\"text-align: center;\">Figure. 3 shows the Seven Layers of Therapeutic Order in Naturopathic Medicine<a href=\"#mla_106\"> [106]</a>.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4acc4f93-18d7-4117-9b62-00dd326b4c0e",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>An Example of a Naturopathic Course of Action</h2>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "For instance, a Naturopathic Doctor (ND) might order a stool test to assess biomarkers of gastrointestinal health. Based on the results, the ND would prescribe nutraceuticals or herbal supplements as part of a baseline protocol to address imbalances in the patient's GI tract. If a patient has experienced rheumatoid arthritis from a young age, with pain stemming from systemic inflammation caused by all bodily systems, the nutraceuticals, and herbal supplements might help correct imbalances such as low stomach acid (hypochlorhydria) caused by a common bacterial infection in the GI tract, H. pylori. The ND might also recommend additional supplements for cleansing and repairing the GI tract to eliminate dysbiosis at later dates <a href=\"#mla_32\"> [32]</a>.\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>What is Traditional Chinese Medicine?</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Traditional Chinese Medicine (TCM) is an ancient practice that spans thousands of years, that has changed little over the centuries. TCM is based on the fundamental concepts of \"Yin-Yang\" and \"Qi\". See Figure. 4 below for a depiction of this fundamental concept of TCM. \"A vital force of life, called Qi, surges through the body. Any imbalance in Qi can cause disease and illness. This imbalance is most commonly thought to be caused by an alteration in the opposite and complementary forces that make up the Qi... called Yin and Yang\" <a href=\"#mla_33\"> [33]</a>.\n",
    "\n",
    "\n",
    "\"The first writings about TCM date back to 200 BCE. Herbal medicine, acupuncture, including theory, practice, diagnosis, and treatment, were recorded in classical Chinese texts and [was] refined over many centuries\" <a href=\"#mla_34\"> [34]</a>.\n",
    "\n",
    "\n",
    "Disease is thought to be alterations in the normal flow of Qi such that Yin and Yang are imbalanced. TCM agrees that an unbalanced Yin and Yang typically has three major causes: external or environmental factors, internal thoughts and emotions, and lifestyle factors.\n",
    "Like Naturopathic Medicine, TCM stimulates the body's own natural restoration and healing processes.\n",
    "\n",
    "\n",
    "Practices used in TCM include:\n",
    "\n",
    "\n",
    "1. Herbal medicine\n",
    "2. Nutrition\n",
    "3. Acupuncture and acupressure\n",
    "4. Moxibustion: Burning a herb near the skin\n",
    "5. Chinese massage: Called Tui Na\n",
    "6. Exercise: Such as Tai Chi and Qi Gong which combine movement with meditation\n",
    "\n",
    "\n",
    "TCM like Naturopathic Medicine also believes that our bodily systems are not individual networks, but rather \"complex networks\"<a href=\"#mla_35\"> [35]</a>. \"According to TCM, Qi flows through organ systems, the kidneys, heart, spleen, liver, lungs, gallbladder, small intestine, and large intestine, by way of meridians\"<a href=\"#mla_36\"> [36]</a> also known as energy pathways.\n",
    "\n",
    "<br><br>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"yinyang.jpg\" alt=\"Fig. 4\" width=\"35%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 4 shows the Fundamental Concept in Traditional Chinese Medicine<a href=\"#mla_107\"> [107]</a>.</p>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>Naturopathic Medicine & Traditional Chinese Medicine Compared</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "\"The term \"naturopathy\" originates from \"natura\" (Latin root for birth) and \"pathos\" (the Greek root for suffering) to suggest \"natural healing\". Naturopaths claim the ancient Greek \"Father of Medicine\", Hippocrates, as the first advocate of Naturopathic Medicine, before the term existed\"<a href=\"#mla_37\"> [37]</a>.\n",
    "\n",
    "Traditional Chinese Medicine stems from \"The Yellow Emperor's Inner Canon (Huangdi Neijing), the oldest received work of Chinese medical theory, [which] was compiled during the Han dynasty around the first century BCE\"<a href=\"#mla_38\"> [38]</a>. \"It was also one of the first books in which the cosmological doctrines of Yinyang...were brought to a mature synthesis\"<a href=\"#mla_39\"> [39]</a>.\n",
    "\n",
    "Naturopathic Medicine like Traditional Chinese Medicine where both founded circa 200 BC. Both ancient practices aim to restore a body's natural ability to heal itself through natural purposes because both practices believe that disease is caused by imbalances within the person. Hippocrates who is the \"Father of Medicine\" postulated that these imbalances primarily stemmed from one's own temperament. The following section aims to describe the vital importance of these temperaments, and how they relate to modern medical theory as a body of my own original work from a lifetime of study. The following paradigm is my life's work which maps temperament and modern medicine together.\n",
    "\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h2>Hippocrates' Humoral Theory, Naturopathic Medicine, Modern Medicine, and Archetypes</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "In medieval and ancient times, all Medical Doctors were instructed in the premise of \"Humoral Theory\" which was founded by Hippocrates <a href=\"#mla_40\"> [40]</a>. Humoral Theory states that all people may be categorized by four different types of temperaments, which are marked by excesses in a specific kind of humor<a href=\"#mla_41\"> [41]</a>. Humor comes from the word human, humor and human from the same root word Humus in Latin which means \"earthly\". To be human is to have excess or deficiencies of certain humors. \"Hippocrates suggested that humors are the vital bodily fluids: blood, phlegm, yellow bile, and black bile<a href=\"#mla_42\"> [42]</a>. Hippocrates postulated that an extreme excess or deficiency of any of the humors (bodily fluid) in a person can be a sign of illness\"<a href=\"#mla_43\"> [43]</a>. Figure. 5 below shows a German 16th-century drawing of the 4 temperments. Hippocrates lived from 460 – c. 370 BC.\n",
    "\n",
    "<br><br>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4temps16th.jpeg\" alt=\"Fig. 5\" width=\"35%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 5 shows a drawing from the 16th-century of The 4 Temperments in Humoral Theory<a href=\"#mla_108\"> [108]</a>.</p>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>The Phlegmatic Temperment</h3>\n",
    "    <hr style=\"width: 45%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsbodycopy.jpeg\" alt=\"Fig. 6\" width=\"15%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 6 shows The Phlegmatic Temperment in Humoral Theory<a href=\"#mla_109\"> [109]</a>.</p>\n",
    "</div>\n",
    "\n",
    "There is much wisdom in making the philosophical claim that excesses and deficiencies, in certain bodily fluids affect your temperament.\n",
    "Take the Phlemagtic temperament. The Phlegmatic is categorized as having an excess bodily fluid of phlegm<a href=\"#mla_44\"> [44]</a>. Having an excess of phlegm is marked by a later scientific discovery known as \"rest and digest\"<a href=\"#mla_45\"> [45]</a>. This is where the parasympathetic system is activated, and all of your bodily energy is focused on resting and digesting food after it is consumed<a href=\"#mla_46\"> [46]</a>. \"Rest and Digest\" is opposed to \"Fight or Flight\"<a href=\"#mla_47\"> [47]</a> which is marked by feeling a sense of danger or preoccupation where someone feels the need to flee a situation or fight a situation that is usually life-threatening<a href=\"#mla_48\"> [48]</a>. The \"Fight or Flight\" state is marked by very high levels of elevated cortisol<a href=\"#mla_49\"> [49]</a>, the release of adrenaline<a href=\"#mla_50\"> [50]</a>, and catecholamines<a href=\"#mla_51\"> [51]</a>, which is what makes the whole \"Fight of Flight\" experience a stressful experience.\n",
    "\n",
    "In \"Rest and Digest\" the digestive system is activated to break down food, so nutrients may be absorbed<a href=\"#mla_52\"> [52]</a>. Conditions known as \"Burnout\" or medically known as \"Adrenal Fatigue\" are marked by chronically low levels of cortisol, which occurs after someone is constantly pathologically being in a state of \"Fight or Flight\"<a href=\"#mla_53\"> [53]</a>.\n",
    "\n",
    "The temperaments go much deeper than just biological processes. Someone who is Phlegmatic, who is constantly in a state of \"Rest and Digest\", is temperamentally marked as someone who is always calm and never stressed. These are people who are very easygoing. But this goes even deeper still. Because someone is temperamentally \"always calm, never stressed, and easy-going\", from an Aristotelian-Thomistic perspective on virtues, they can be prone to laziness <a href=\"#mla_54\"> [54]</a>. Aristotelian-Thomistic virtue ethics states all temperaments have inherent strengths and weaknesses<a href=\"#mla_55\"> [55]</a>. \n",
    "\n",
    "Someone who is Phlegmatic, who is constantly in a state of \"Rest and Digest\" is said to have a high concupiscible appetite, which is defined as someone \"Who has a great appetite to move towards many pleasurable inherent goods, and move away from arduous evils\"<a href=\"#mla_56\"> [56]</a>. As such, this temperament is naturally very stoic, calm, and peaceful but they may be prone to laziness and be less dutiful from an Aristotelian-Thomistic perspective on virtue ethics that are dispositions in someone's character<a href=\"#mla_57\"> [57]</a>.\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>The Choleric Temperment</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsbodycopy2.jpeg\" alt=\"Fig. 7\" width=\"15%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 7 shows The Choleric Temperment in Humoral Theory<a href=\"#mla_110\"> [110]</a>.</p>\n",
    "</div>\n",
    "\n",
    "The Phlegmatic temperament is as opposed to the Choleric Temperament, which is marked by a high irascible appetite which is defined as \"a great appetite to move towards arduous inherent goods, and away from easy evils\"<a href=\"#mla_58\"> [58]</a>. Hippocrates in Humoral Theory marked the Choleric Temperament as someone who has an excess of the humor \"yellow bile\"<a href=\"#mla_59\"> [59]</a>. It was later discovered by modern science that yellow bile is a product that is stored in the gallbladder of a person, that sits just under the liver <a href=\"#mla_60\"> [60]</a>. Yellow bile is a product of detoxed toxins from stage II detoxification pathways before they are used one more time in the digestive system to help break down fats which are the most calorically dense macronutrient you can eat<a href=\"#mla_61\"> [61]</a>.\n",
    "\n",
    "The Choleric temperament goes deeper. Since \"yellow bile\" is the result of detoxed toxins that an individual interfaces with on a daily basis, this temperament is defined as someone who is a \"leader\", \"very hard worker\", and \"spearhead of work\"<a href=\"#mla_62\"> [62]</a>. Hence, since so much of this temperament aligns with \"doing\", this temperament will inevitably absorb more toxins through the skin, and lungs, which inevitably hits our livers to be dealt with in a detoxification process<a href=\"#mla_63\"> [63]</a>.\n",
    "\n",
    "Someone who is a Choleric, who is constantly working to arduous goods, may be more prone to anger hence the word Choleric which comes from the Latin word for anger \"Cholera\"<a href=\"#mla_64\"> [64]</a>.  They are typically less tolerant of people and their mistakes from an Aristotelian-Thomistic perspective on virtue ethics within someone's character<a href=\"#mla_65\"> [65]</a>. They normally deal with more stress than all the other temperaments, so they may be more easily prone to adrenal fatigue or \"burnout\"<a href=\"#mla_66\"> [66]</a><a href=\"#mla_67\"> [67]</a>. There is much wisdom from Hippocrates' school of thought and his Humoral Theory indeed.\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>The Melancholic Temperment</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsbodycopy3.jpeg\" alt=\"Fig. 8\" width=\"15%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 8 shows The Melancholic Temperment in Humoral Theory<a href=\"#mla_111\"> [111]</a>.</p>\n",
    "</div>\n",
    "\n",
    "The Melancholic temperament is on a different axis than what the Phelgmatics or Cholerics are on. The Phlegmatic is to what is to \"be\" or \"rest\", where you do not have to think, feel, or do anything to \"be\". You just are the summation of everything that you think, feel, or do.\n",
    "Choleric is to \"doing\", it is an active principle of human nature.\n",
    "\n",
    "Melancholics are on the thinking and feeling axis, whereas Melancholics are to \"reflection\", and \"discovering and believing the truth for the sake of truth\"<a href=\"#mla_68\"> [68]</a>, the final temperament Sanguine is to \"feeling\" and \" feeling interconnected with everything\"<a href=\"#mla_69\"> [69]</a>.\n",
    "\n",
    "Hippocates stated that Melancholics have an excess of \"black bile\" which was a liquid that was found in the spleen at the time of Hippocates<a href=\"#mla_70\"> [70]</a>. \"Black Bile was thought to originate in the spleen, meaning that melancholy... [was]...blamed on the spleen. With an overproduction of Black Bile, a person may have appeared cold and dry, sad in disposition, and overall doubtful of the world\"<a href=\"#mla_71\"> [71]</a>. Modern medicine states that the spleen \"stores and filters blood and makes white blood cells that protect you from infection\"<a href=\"#mla_72\"> [72]</a> and \"the spleen is part of the lymphatic system\" <a href=\"#mla_73\"> [73]</a>. The lymphatic system deals with the waste that is naturally left behind from natural blood circulation within blood plasma<a href=\"#mla_74\"> [74]</a>.\n",
    "\n",
    "\n",
    "Since fluid from the spleen is a concentration of white blood cells and lymph, modern medicine affirms that excesses in spleen fluid mark someone who biologically produces more waste in their natural blood circulation<a href=\"#mla_75\"> [75]</a>, hence the reflective nature of this temperament, because \"doing\" things, and their actions that follow, may lead to greater waste that is produced by their blood exertion in \"doing things\", which is then picked up by the lymphatic system in blood plasma<a href=\"#mla_76\"> [76]</a>, which connects to the spleen<a href=\"#mla_77\"> [77]</a>.\n",
    "\n",
    "\n",
    "\n",
    "Someone who is a Melancholic, who is constantly thinking to understand truths, may be more prone to melancholy hence the word Melancholic which comes from the Latin word for sadness \"Melancholicus\".  They typically react more slowly and are affected for a long time from negative events, in an Aristotelian-Thomistic perspective on virtue ethics within someone's character<a href=\"#mla_78\"> [78]</a>. Aristotelian-Thomistic philosophy states that this stems from this temperament of \"never wanting to suffer\"<a href=\"#mla_79\"> [79]</a> hence them falling into a character defect known as fear<a href=\"#mla_80\"> [80]</a>. Fear is defined as \"the feeling that an inevitable evil is coming your way, that you cannot stop\"<a href=\"#mla_81\"> [81]</a> which is why this temperament is by nature also very reflective because they do not want to suffer negative things<a href=\"#mla_82\"> [82]</a>. This can be offset by growing in the virtues of hope, joy, and gratefulness for one's state in life<a href=\"#mla_83\"> [83]</a>.\n",
    "\n",
    "Melancholics normally deal with higher levels of serotonin than all the other temperaments, which is an inhibitory neurotransmitter, they are inhibitory towards making rash decisions. Modern Medicine states serotonin and dopamine work in a push-pull fashion, when one goes lower the other tends to go higher<a href=\"#mla_84\"> [84]</a>. A Sanguine typically has high constant levels of dopamine, and thus lower levels of serotonin. Dopamine is an excitatory feel-good neurotransmitter for \"pursuing goals\"<a href=\"#mla_85\"> [85]</a>, and serotonin is the inhibitory feel-good neurotransmitter for \"feeling content with what you have\"<a href=\"#mla_86\"> [86]</a>. Melancholics are more easily prone to melancholy or \"anxiety\" because anxiety is marked by not having enough dopamine to \"chase rewards\" as a precursor to adrenaline that is released in perceived stressful situations in an anxiety disorder<a href=\"#mla_87\"> [87]</a>. There is much wisdom from Hippocrates' School of Thought and his Humoral Theory, which still relates to Modern Medicine today.\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>The Sanguine Temperment</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsbodycopy4.jpeg\" alt=\"Fig. 9\" width=\"15%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 9 shows The Sanguine Temperment in Humoral Theory<a href=\"#mla_112\"> [112]</a>.</p>\n",
    "</div>\n",
    "\n",
    "The final temperament of the four temperaments is the Sanguine Temperament. It is on the same axis as the melancholic temperament. Melancholic is to thinking, as Sanguine is to feeling. The Phlegmatic is to \"being\", and Choleric is to \"doing\".\n",
    "\n",
    "Hippocrates stated that Sanguine's have an excess of the blood fluid<a href=\"#mla_88\"> [88]</a>. Sanguines are defined as \"very charismatic people\", they are very social, they want to be connected to everything and they want to feel everything, hence the flow the fluid from the heart<a href=\"#mla_89\"> [89]</a>. Heart for feeling, blood for passion, and whole body interconnectedness. \n",
    "\n",
    "Philosophically, the Sanguine differs from the Melancholic in that Sanguines do not have to deal with excess waste within the plasma in their natural blood circulation, which the Melancholic's spleen and lymph have to deal with. The Melanchonics have higher rates of waste in their blood circulation by nature, and the Sanguine does not. The Sanguines philosophically deal with less circulating waste in their spleen and plasma, hence why they want to \"feel everything\"<a href=\"#mla_90\"> [90]</a> and be \"connected to everything in nature\"<a href=\"#mla_91\"> [91]</a>, to have a sense of \"oneness\"<a href=\"#mla_92\"> [92]</a> that starts in the ventricles of their hearts.\n",
    "\n",
    "Sanguines are defined from an Aristotelian-Thomistic perspective as \"reacting very quickly to negative events, and being undisturbed by them\"<a href=\"#mla_93\"> [93]</a>, which is the exact opposite of a Melancholic who \"reacts very slowly to negative events, and is affected for a long time by them\"<a href=\"#mla_94\"> [94]</a>.\n",
    "\n",
    "\n",
    "Someone who is a Sanguine who reacts very quickly to negative events, remains undisturbed by them, is very social and very charismatic is from an Aristotelian-Thomistic view \"someone who is prone to intemperance\" from the dispositions within their character regarding their virtues<a href=\"#mla_95\"> [95]</a>.\n",
    "Aristotelian-Thomistic philosophy states why this stems from this temperament, it is because by nature they tend to \"never want to be apart from pleasure\"<a href=\"#mla_96\"> [96]</a> hence them falling into intemperances such as drug addiction and alcoholism, in which alcoholism is found to have a genetic predisposition within families<a href=\"#mla_97\"> [97]</a> that are Sanguine. This points to the Modern Medical concept of Epigenetics which states in general that \"The summation of all your experiences, good and bad, all influence your genetics, and the genes that are passed down to the next generation up until the very moment of conception of a newborn\"<a href=\"#mla_98\"> [98]</a>.\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>Beyond Just Temperments</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsfaces.jpeg\" alt=\"Fig. 10\" width=\"35%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 10 shows The Faces of the 4 Temperments in Humoral Theory<a href=\"#mla_113\"> [113]</a>.</p>\n",
    "</div>\n",
    "\n",
    "The paradigm goes much deeper than just temperaments, natural virtue strengths and weaknesses, and biological disease. There is a body of work published by Alexander Lowen who was a student of Wilhelm Reich, a psychologist who was a student of Sigmund Freud, describes in his work that one's temperament and character have \"influence in how your anatomy and physiology develops, and in turn, even looks like once adulthood is reached\"<a href=\"#mla_99\"> [99]</a>. That is to say that one's virtues and defects (or character excesses and defects), which symbiotically pair with one's blend of temperments from Hippocrates' Humoral Theory, determine one's own psychology and how one may physically appear in person. \n",
    "\n",
    "Wilhelm Reich was a direct student of Sigmund Freud<a href=\"#mla_100\"> [100]</a>, Freud pioneered modern Psychology<a href=\"#mla_101\"> [101]</a>. Reich's \"work on character and the idea of muscular armoring contributed to the development of what is now known as... bioenergetic analysis of Reich's student Alexander Lowen\"<a href=\"#mla_102\"> [102]</a> who described formations of human anatomy and physiology as their own kind of \"archetype\"<a href=\"#mla_103\"> [103]</a> which is analogous to Hippocrates' Humoral Theory which describes the four temperaments as their own archetypes.  Hippocrates's Humoral Theory and Lowen's Bioenergetic Theory are analogous, in that in my own research the temperament archetypes and the \"character structure\" or \"physical appearance\" archetypes overlap in their similarities. In other words, each psychological temperament (Humoral Temperament) corresponds to a person's physical appearance in many regards. The body is the mind, and the mind is the body in many respects.\n",
    "\n",
    "These two sets of archetypes may map to each other, but they are not \"hard and fast rules\" they are not \"completely definitive 100% of the time\". It is a metaphysical framework for understanding nature and illnesses. Despite my own life's work in researching and mapping these two paradigms, with the paradigm of modern science, they still describe patterns that can be found within human populations until this day. Patterns are not 100% completely definitive as to what a person's temperament and bioenergetic character structure might actually be, we are usually a blend of two temperaments. These two paradigms only describe archetypes of the human condition, which has been seen since ancient Greek times, and one may postulate that the human condition will not change anytime in the near future. This metaphysical framework still has medical and personal value. Regarding the linking of Hippocrates' Humoral Theory and Lowen's Bioenergetic Analysis, there is much truth to this based on a fundamental concept of biology that states \"form follows function\"<a href=\"#mla_104\"> [104]</a> which is from Darwin's Evolution Theory. The nature of nature is all a pseudo-science at its most fundamental concepts, and this is a whole study and discipline that is outside the scope of being further elaborated in this body of work.\n",
    "\n",
    "<br><br>\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <img src=\"4tempsbody.jpeg\" alt=\"Fig. 11\" width=\"55%\">\n",
    "    <br><br>\n",
    "    <p style=\"text-align: center;\">Figure. 11 shows all 4 Temperments in Humoral Theory<a href=\"#mla_114\"> [114]</a>.</p>\n",
    "</div>\n",
    "\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>The Scope of this Body of Work</h3>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>\n",
    "\n",
    "The scope of this body of work is limited to the first and final layer (sixth layer) of the \"Bodily Health Systems: Their Sequential Importance and Interconnection\" section as seen above. This project aims to draw inferences from four datasets which are research studies, that measure biomarkers in genetics, gut microbiota, and biomarkers as seen in the pathology of Alzheimer's disease in the different stages of progression of this pathological phenomenon. Several studies measure the effects of herbal medicine from Naturopathic Medicine and Traditional Chinese Medicine Botany in these two layers of bodily health systems. The purpose of the analysis of this body of work, is to draw inferences from herbal medicine, its effects on gut microbiota, colonic and overall gene expression, and whether or not it has a positive or negative influence on Alzheimer's pathology for research purposes that aim to answer this body of work's project proposal as seen in the section below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fb49d50-852b-4a10-820c-372f2179db68",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>1.1.2 Proposal of Capstone</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "How does Naturopathic Medicine and Traditional Chinese Medicine herbs, such as Danggui Shaoyao San and Rhodiola Rosea, affect the gut microbiome and thus colonic and overall gene expression, to reduce or prevent the progression of Alzheimer’s disease and its pathologies?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bf7b71-c3f1-468c-82a8-0dc3049a8140",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br><br>\n",
    "    <h3>1.1.3 Criteria for Success of Capstone</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "Successful criteria shall include…\n",
    "- 1.) Mapping the effects of Danggui Shaoyao San on the gut microbiome.\n",
    "- 2.) Mapping the effects of Rhodiola Rosea, on the gut microbiome.\n",
    "- 3.) Mapping the effects of Danggui Shaoyao San on colonic and overall gene expression.\n",
    "- 4.) Mapping the effects of Rhodiola Rosea on colonic and overall gene expression.\n",
    "- 5.) Mapping the relationship between Danggui Shaoyao, the gut microbiome, colonic, and overall gene expression, and whether it reduces or prevents the progression of Alzheimer’s disease and its pathologies.\n",
    "- 6.) Mapping the relationship between Rhodiola Rosea, the gut microbiome, colonic, and overall gene expression, and whether it reduces or prevents the progression of Alzheimer’s disease and its pathologies.\n",
    "- 7.) Determining which herbal supplement is more efficacious at preventing or reducing Alzheimer’s disease and its pathologies.\n",
    "- 8.) Determining a recommended course of action with these two supplements, and how Naturopathic Medicine and Traditional Chinese Medicine fit as a whole into the inferences of the proposal statement."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ae0465-6e06-4f90-b4ae-30c156df2b60",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h2>1.2 Imports</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br><br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4b966f0b-1660-475e-9984-f6b72c71c8c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "python(20500) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: missingno in ./anaconda3/lib/python3.11/site-packages (0.5.2)\n",
      "Requirement already satisfied: numpy in ./anaconda3/lib/python3.11/site-packages (from missingno) (1.23.5)\n",
      "Requirement already satisfied: matplotlib in ./anaconda3/lib/python3.11/site-packages (from missingno) (3.8.4)\n",
      "Requirement already satisfied: scipy in ./anaconda3/lib/python3.11/site-packages (from missingno) (1.13.1)\n",
      "Requirement already satisfied: seaborn in ./anaconda3/lib/python3.11/site-packages (from missingno) (0.13.2)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (1.4.4)\n",
      "Requirement already satisfied: packaging>=20.0 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (3.0.9)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in ./anaconda3/lib/python3.11/site-packages (from matplotlib->missingno) (2.9.0.post0)\n",
      "Requirement already satisfied: pandas>=1.2 in ./anaconda3/lib/python3.11/site-packages (from seaborn->missingno) (2.2.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in ./anaconda3/lib/python3.11/site-packages (from pandas>=1.2->seaborn->missingno) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in ./anaconda3/lib/python3.11/site-packages (from pandas>=1.2->seaborn->missingno) (2023.3)\n",
      "Requirement already satisfied: six>=1.5 in ./anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib->missingno) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "# Import standard libraries and packages\n",
    "\n",
    "!pip install missingno \n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import missingno as msno\n",
    "import io as io\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6dbfa0a5-c02d-46d6-9a83-d4f6a97fd41d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------*HEAD*---------------------\n",
      "  function                                  description     DSS_6     DSS_5  \\\n",
      "0  COG0001    Glutamate-1-semialdehyde aminotransferase  28746.42  25319.50   \n",
      "1  COG0002   N-acetyl-gamma-glutamylphosphate reductase  47780.21  35611.67   \n",
      "2  COG0003  Anion-transporting ATPase, ArsA/GET3 family    580.28    650.88   \n",
      "3  COG0004                 Ammonia channel protein AmtB  52589.01  28377.26   \n",
      "4  COG0005              Purine nucleoside phosphorylase  51522.40  27235.22   \n",
      "\n",
      "      DSS_4     DSS_3     DSS_2     DSS_1       M_5       M_4       M_3  \\\n",
      "0  38659.43   8318.37   9410.15  32813.32  52762.65  30121.00  13857.36   \n",
      "1  51824.55   9909.72  19836.70  48508.07  62180.41  48010.13  30178.47   \n",
      "2    800.90    870.28    413.26    689.30   2206.97    902.00   1393.69   \n",
      "3  61140.95  61382.22  24997.03  56614.41  59554.28  50161.83  33047.73   \n",
      "4  49603.71  15641.12  17534.30  42615.71  65180.88  41739.31  21916.03   \n",
      "\n",
      "        M_2       M_1     Con_4     Con_3     Con_2     Con_1  \n",
      "0  35318.12  26756.76  12545.71  18077.38  12354.19  12171.57  \n",
      "1  49923.14  42734.80  26232.08  33070.83  24953.04  40474.55  \n",
      "2   1078.46   1333.87    763.66   1052.44   1303.08    772.10  \n",
      "3  54770.38  44917.57  42383.58  48793.17  45455.87  41664.39  \n",
      "4  45843.87  35995.22  22046.37  27221.00  16584.69  17016.15  \n",
      "-------------------*END HEAD*-------------------\n"
     ]
    }
   ],
   "source": [
    "# Import Dataset 1 : xlsx file\n",
    "# (Danggui Shaoyao San: A comprehensive modulation of the microbiota-gut-brain axis for attenuating Alzheimer’s disease-related pathology)\n",
    "\n",
    "file_path_1 = '/Volumes/Extreme SSD/Recent Files/Capstone Datasets/Table5_Danggui%20Shaoyao%20San%3A%20comprehensive%20modulation%20of%20the%20microbiota-gut-brain%20axis%20for%20attenuating%20Alzheimer%3Fs%20disease-related%20pathology.XLSX'\n",
    "\n",
    "# Load the Excel file into a pandas DataFrame\n",
    "ds_1 = pd.read_excel(file_path_1)\n",
    "\n",
    "# Display the first few rows of the DataFrame to verify the import\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_1.head())\n",
    "print('-------------------*END HEAD*-------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "97f3fd8d-ff4a-4461-877c-68a8d60a1031",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------*HEAD*---------------------\n",
      "  Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "0                                                NaN                                               \n",
      "1                                            Samples                                               \n",
      "2                                          GSM300167                                               \n",
      "3                                          GSM300168                                               \n",
      "4                                          GSM300169                                               \n",
      "\n",
      "               Unnamed: 1 Unnamed: 2   Unnamed: 3 Unnamed: 4 Unnamed: 5  \\\n",
      "0                     NaN        NaN          NaN        NaN        NaN   \n",
      "1            Brain region     Gender  Braak stage      ATP5B       BDNF   \n",
      "2  superior frontal gyrus     female            0   0.671874    0.80707   \n",
      "3             hippocampus     female            0   0.972526   2.671879   \n",
      "4             hippocampus       male            0   0.744222    3.33928   \n",
      "\n",
      "  Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14  \\\n",
      "0        NaN        NaN        NaN        NaN  ...         NaN         NaN   \n",
      "1       CD44        DLD     GABRG2       GAD2  ...      PSMD14     SLC32A1   \n",
      "2   1.981505   0.732005    0.32877   0.641998  ...    0.681275    0.165948   \n",
      "3   7.455322   1.066505   0.640967   0.352101  ...    1.247155    1.211095   \n",
      "4   3.321804   0.877786   0.357565   0.317235  ...    1.120629    0.621519   \n",
      "\n",
      "  Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \\\n",
      "0         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "1       SMAD4      SNAP25         SYP        SYT1        TJP1       VDAC1   \n",
      "2    0.857919    0.914064    0.643572    0.685069    1.045188    0.557694   \n",
      "3    0.734436    0.752652    0.912192    0.710386    0.765581    0.845077   \n",
      "4     0.96861    0.690655    0.757741    0.564051    0.904162    0.558936   \n",
      "\n",
      "  Unnamed: 21 Unnamed: 22  \n",
      "0         NaN         NaN  \n",
      "1       WWTR1        YAP1  \n",
      "2    1.467627    1.755059  \n",
      "3    1.629639    1.158677  \n",
      "4    1.518328    1.704543  \n",
      "\n",
      "[5 rows x 23 columns]\n",
      "-------------------*END HEAD*-------------------\n"
     ]
    }
   ],
   "source": [
    "# Import Dataset 2 : xlsx file\n",
    "# (Exploring the Key Genes and Identification of Potential Diagnosis Biomarkers in Alzheimer’s Disease Using Bioinformatics Analysis)\n",
    "\n",
    "file_path_2 = '/Volumes/Extreme SSD/Recent Files/Capstone Datasets/Table_4_Exploring the Key Genes and Identification of Potential Diagnosis Biomarkers in AlzheimerÃ¢ÂÂs Disease Using Bioinformatics Analysis.XLSX'\n",
    "\n",
    "# Load the Excel file into a pandas DataFrame\n",
    "ds_2 = pd.read_excel(file_path_2)\n",
    "\n",
    "# Display the first few rows of the DataFrame to verify the import\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_2.head())\n",
    "print('-------------------*END HEAD*-------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "18c5305a-d280-46b6-afb5-a157402ecf84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------*HEAD*---------------------\n",
      "  Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A\n",
      "0    83         W   F          34.75           NaN  NaN\n",
      "1    85         W   F          25.57     29.130000  NaN\n",
      "2    86         W   M          33.91     37.609234  NaN\n",
      "3    87         W   M      37.272731     38.420000  NaN\n",
      "4    89         W   M          28.93     29.660000  NaN\n",
      "-------------------*END HEAD*-------------------\n"
     ]
    }
   ],
   "source": [
    "# Import Dataset 3 : xlsx/xlsm files\n",
    "# (The impact of Rhodiola Rosea on biomarkers of diabetes, inflammation, and microbiota in a leptin receptor-knockout mouse model)\n",
    "\n",
    "# This dataset compares a gene of interest, versus a control gene within the liver of mice to asses the efficacy of gene expression \n",
    "# For mice that take Rhodiola Rosea vesus the control group which is just water\n",
    "\n",
    "file_path_3 = '/Volumes/Extreme SSD/Recent Files/Capstone Datasets/doi_10_7280_D1FX2D__v20220616/Liver_CRP_qPCR_Data.xlsm'\n",
    "\n",
    "# Load the Excel file into a pandas DataFrame\n",
    "ds_3 = pd.read_excel(file_path_3)\n",
    "\n",
    "# Display the first few rows of the DataFrame to verify the import\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_3.head())\n",
    "print('-------------------*END HEAD*-------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8533cfbd-e693-441e-ac74-3533f5e5eea3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------*HEAD*---------------------\n",
      "     ID_REF  GSM3101204  GSM3101205  GSM3101206  GSM3101207  GSM3101208  \\\n",
      "0  10338001   11.621818   11.723674   11.797580   11.845637   11.706340   \n",
      "1  10338002    4.882116    4.819835    4.945165    4.816617    5.066188   \n",
      "2  10338003    9.815821    9.913645   10.070279   10.068832    9.953928   \n",
      "3  10338004    9.123786    9.135206    9.285084    9.463849    9.290599   \n",
      "4  10338005    2.294406    2.362550    2.367593    2.393940    2.387859   \n",
      "\n",
      "   GSM3101209  GSM3101210  GSM3101211  GSM3101212  ...  GSM3101220  \\\n",
      "0   11.619102   11.640126   11.753207   11.671719  ...   11.662705   \n",
      "1    5.032585    4.809279    4.929813    4.989135  ...    4.954034   \n",
      "2    9.841130    9.849367    9.990176    9.907970  ...    9.884320   \n",
      "3    9.101129    9.287354    9.266923    9.084664  ...    9.149640   \n",
      "4    2.398327    2.437759    2.382819    2.403590  ...    2.363132   \n",
      "\n",
      "   GSM3101221  GSM3101222  GSM3101223  GSM3101224  GSM3101225  GSM3101226  \\\n",
      "0   11.777588   11.689259   11.636659   11.690642   11.611120   11.738238   \n",
      "1    4.783552    4.968062    5.006434    4.920746    4.877134    4.843854   \n",
      "2    9.921198    9.963101    9.855835    9.898591    9.785410    9.955079   \n",
      "3    9.297222    9.093457    9.064904    8.989178    9.049290    9.150534   \n",
      "4    2.345471    2.368008    2.357114    2.355657    2.301924    2.357530   \n",
      "\n",
      "   GSM3101227  GSM3101228  GSM3101229  \n",
      "0   11.653844   11.635455   11.663165  \n",
      "1    4.951407    4.874028    4.746962  \n",
      "2    9.833608    9.770072    9.802227  \n",
      "3    9.144118    8.987364    9.074741  \n",
      "4    2.324505    2.324368    2.361027  \n",
      "\n",
      "[5 rows x 27 columns]\n",
      "-------------------*END HEAD*-------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yt/sy7z4yq9639_05r46b0x1gmm0000gn/T/ipykernel_72595/3970757467.py:29: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ds_4 = pd.read_csv(io.StringIO(data_str), delimiter='\\t', header=0)\n"
     ]
    }
   ],
   "source": [
    "# Import Dataset 4 : txt file\n",
    "# (Integrative analysis of gut microbiota composition, host colonic gene expression and intraluminal metabolites in aging C57BL/6J mice)\n",
    "\n",
    "# This dataset compares mice at various ages, and their colonic gene expressions that resulted from their sampled microbiota\n",
    "\n",
    "file_path_4 = '/Volumes/Extreme SSD/Recent Files/Capstone Datasets/GSE113257_family.xml/TXT FILE OF DATA/GSE113257_series_matrix.txt'\n",
    "\n",
    "# Read the file into a list of lines\n",
    "with open(file_path_4, 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Identify where the data section starts\n",
    "data_start_index = None\n",
    "for i, line in enumerate(lines):\n",
    "    if line.startswith('!series_matrix_table_begin'):\n",
    "        data_start_index = i + 1  # The data starts on the line after this header\n",
    "        break\n",
    "\n",
    "# Check if data_start_index was found\n",
    "if data_start_index is None:\n",
    "    raise ValueError(\"Data section header '!series_matrix_table_begin' not found.\")\n",
    "\n",
    "# Extract header lines and data lines\n",
    "header_lines = lines[:data_start_index - 1]\n",
    "data_lines = lines[data_start_index:]\n",
    "\n",
    "# Process the data lines into a DataFrame\n",
    "data_str = \"\\n\".join(data_lines)\n",
    "ds_4 = pd.read_csv(io.StringIO(data_str), delimiter='\\t', header=0)\n",
    "\n",
    "# Print first few rows\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_4.head())\n",
    "print('-------------------*END HEAD*-------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0b764022-89c3-49ad-ac5b-9ea0e27854bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------*DATA TYPES*------------------\n",
      "ID_REF         object\n",
      "GSM3101204    float64\n",
      "GSM3101205    float64\n",
      "GSM3101206    float64\n",
      "GSM3101207    float64\n",
      "GSM3101208    float64\n",
      "GSM3101209    float64\n",
      "GSM3101210    float64\n",
      "GSM3101211    float64\n",
      "GSM3101212    float64\n",
      "GSM3101213    float64\n",
      "GSM3101214    float64\n",
      "GSM3101215    float64\n",
      "GSM3101216    float64\n",
      "GSM3101217    float64\n",
      "GSM3101218    float64\n",
      "GSM3101219    float64\n",
      "GSM3101220    float64\n",
      "GSM3101221    float64\n",
      "GSM3101222    float64\n",
      "GSM3101223    float64\n",
      "GSM3101224    float64\n",
      "GSM3101225    float64\n",
      "GSM3101226    float64\n",
      "GSM3101227    float64\n",
      "GSM3101228    float64\n",
      "GSM3101229    float64\n",
      "dtype: object\n",
      "----------------*END DATA TYPES*----------------\n"
     ]
    }
   ],
   "source": [
    "# Print data types to check mixed types\n",
    "print('------------------*DATA TYPES*------------------')\n",
    "print(ds_4.dtypes)\n",
    "print('----------------*END DATA TYPES*----------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a95ebbed-5dc8-40c5-86cc-ff8668496f81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------*HEAD*---------------------\n",
      "     ID_REF  GSM3101204  GSM3101205  GSM3101206  GSM3101207  GSM3101208  \\\n",
      "0  10338001   11.621818   11.723674   11.797580   11.845637   11.706340   \n",
      "1  10338002    4.882116    4.819835    4.945165    4.816617    5.066188   \n",
      "2  10338003    9.815821    9.913645   10.070279   10.068832    9.953928   \n",
      "3  10338004    9.123786    9.135206    9.285084    9.463849    9.290599   \n",
      "4  10338005    2.294406    2.362550    2.367593    2.393940    2.387859   \n",
      "\n",
      "   GSM3101209  GSM3101210  GSM3101211  GSM3101212  ...  GSM3101220  \\\n",
      "0   11.619102   11.640126   11.753207   11.671719  ...   11.662705   \n",
      "1    5.032585    4.809279    4.929813    4.989135  ...    4.954034   \n",
      "2    9.841130    9.849367    9.990176    9.907970  ...    9.884320   \n",
      "3    9.101129    9.287354    9.266923    9.084664  ...    9.149640   \n",
      "4    2.398327    2.437759    2.382819    2.403590  ...    2.363132   \n",
      "\n",
      "   GSM3101221  GSM3101222  GSM3101223  GSM3101224  GSM3101225  GSM3101226  \\\n",
      "0   11.777588   11.689259   11.636659   11.690642   11.611120   11.738238   \n",
      "1    4.783552    4.968062    5.006434    4.920746    4.877134    4.843854   \n",
      "2    9.921198    9.963101    9.855835    9.898591    9.785410    9.955079   \n",
      "3    9.297222    9.093457    9.064904    8.989178    9.049290    9.150534   \n",
      "4    2.345471    2.368008    2.357114    2.355657    2.301924    2.357530   \n",
      "\n",
      "   GSM3101227  GSM3101228  GSM3101229  \n",
      "0   11.653844   11.635455   11.663165  \n",
      "1    4.951407    4.874028    4.746962  \n",
      "2    9.833608    9.770072    9.802227  \n",
      "3    9.144118    8.987364    9.074741  \n",
      "4    2.324505    2.324368    2.361027  \n",
      "\n",
      "[5 rows x 27 columns]\n",
      "-------------------*END HEAD*-------------------\n"
     ]
    }
   ],
   "source": [
    "# The warning is for mixed data types for 'ID_REF'\n",
    "# The following code changes this\n",
    "\n",
    "# Convert ID_REF to string type, becuase column headers are strings\n",
    "ds_4['ID_REF'] = ds_4['ID_REF'].astype(str)\n",
    "\n",
    "# Let's check again to see if the warning is gone\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_4.head())\n",
    "print('-------------------*END HEAD*-------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1114aac5-305d-4145-a32a-948dab7c9ddf",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h2>1.3 Data Definition</h2>\n",
    "    <hr style=\"width: 100%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01c12c70-a54e-4fcd-956b-1e2c734d6c6c",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.1 Dataset 1</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f0bbe9f-5f72-4685-b2e4-4834f3ec150e",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.2 Dataset One Overview</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b657007e-a82b-4da1-83e5-b87474d17aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 1*----------\n",
      "--------------------*INFO*----------------------\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4108 entries, 0 to 4107\n",
      "Data columns (total 17 columns):\n",
      " #   Column       Non-Null Count  Dtype  \n",
      "---  ------       --------------  -----  \n",
      " 0   function     4108 non-null   object \n",
      " 1   description  4108 non-null   object \n",
      " 2   DSS_6        4108 non-null   float64\n",
      " 3   DSS_5        4108 non-null   float64\n",
      " 4   DSS_4        4108 non-null   float64\n",
      " 5   DSS_3        4108 non-null   float64\n",
      " 6   DSS_2        4108 non-null   float64\n",
      " 7   DSS_1        4108 non-null   float64\n",
      " 8   M_5          4108 non-null   float64\n",
      " 9   M_4          4108 non-null   float64\n",
      " 10  M_3          4108 non-null   float64\n",
      " 11  M_2          4108 non-null   float64\n",
      " 12  M_1          4108 non-null   float64\n",
      " 13  Con_4        4108 non-null   float64\n",
      " 14  Con_3        4108 non-null   float64\n",
      " 15  Con_2        4108 non-null   float64\n",
      " 16  Con_1        4108 non-null   float64\n",
      "dtypes: float64(15), object(2)\n",
      "memory usage: 545.7+ KB\n",
      "None\n",
      "------------------*END INFO*--------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*COLUMNS*--------------------\n",
      "Index(['function', 'description', 'DSS_6', 'DSS_5', 'DSS_4', 'DSS_3', 'DSS_2',\n",
      "       'DSS_1', 'M_5', 'M_4', 'M_3', 'M_2', 'M_1', 'Con_4', 'Con_3', 'Con_2',\n",
      "       'Con_1'],\n",
      "      dtype='object')\n",
      "-----------------*END COLUMNS*------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------------------*DATA TYPES*------------------\n",
      "function        object\n",
      "description     object\n",
      "DSS_6          float64\n",
      "DSS_5          float64\n",
      "DSS_4          float64\n",
      "DSS_3          float64\n",
      "DSS_2          float64\n",
      "DSS_1          float64\n",
      "M_5            float64\n",
      "M_4            float64\n",
      "M_3            float64\n",
      "M_2            float64\n",
      "M_1            float64\n",
      "Con_4          float64\n",
      "Con_3          float64\n",
      "Con_2          float64\n",
      "Con_1          float64\n",
      "dtype: object\n",
      "----------------*END DATA TYPES*----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "---------------------*HEAD*---------------------\n",
      "  function                                  description     DSS_6     DSS_5  \\\n",
      "0  COG0001    Glutamate-1-semialdehyde aminotransferase  28746.42  25319.50   \n",
      "1  COG0002   N-acetyl-gamma-glutamylphosphate reductase  47780.21  35611.67   \n",
      "2  COG0003  Anion-transporting ATPase, ArsA/GET3 family    580.28    650.88   \n",
      "3  COG0004                 Ammonia channel protein AmtB  52589.01  28377.26   \n",
      "4  COG0005              Purine nucleoside phosphorylase  51522.40  27235.22   \n",
      "\n",
      "      DSS_4     DSS_3     DSS_2     DSS_1       M_5       M_4       M_3  \\\n",
      "0  38659.43   8318.37   9410.15  32813.32  52762.65  30121.00  13857.36   \n",
      "1  51824.55   9909.72  19836.70  48508.07  62180.41  48010.13  30178.47   \n",
      "2    800.90    870.28    413.26    689.30   2206.97    902.00   1393.69   \n",
      "3  61140.95  61382.22  24997.03  56614.41  59554.28  50161.83  33047.73   \n",
      "4  49603.71  15641.12  17534.30  42615.71  65180.88  41739.31  21916.03   \n",
      "\n",
      "        M_2       M_1     Con_4     Con_3     Con_2     Con_1  \n",
      "0  35318.12  26756.76  12545.71  18077.38  12354.19  12171.57  \n",
      "1  49923.14  42734.80  26232.08  33070.83  24953.04  40474.55  \n",
      "2   1078.46   1333.87    763.66   1052.44   1303.08    772.10  \n",
      "3  54770.38  44917.57  42383.58  48793.17  45455.87  41664.39  \n",
      "4  45843.87  35995.22  22046.37  27221.00  16584.69  17016.15  \n",
      "-------------------*END HEAD*-------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*DESCRIBE*-------------------\n",
      "               DSS_6          DSS_5          DSS_4         DSS_3  \\\n",
      "count    4108.000000    4108.000000    4108.000000    4108.00000   \n",
      "mean    20648.337578   18468.828065   27138.905791   23587.57852   \n",
      "std     38239.891142   27214.156844   42532.202252   46501.93423   \n",
      "min         0.000000       0.000000       0.000000       0.00000   \n",
      "25%       171.390000     403.882500     490.410000     158.17000   \n",
      "50%      2613.945000    7752.165000   12103.535000    2572.94500   \n",
      "75%     32356.337500   26449.615000   39585.192500   49243.32750   \n",
      "max    472794.090000  291903.080000  528548.530000  613861.45000   \n",
      "\n",
      "               DSS_2          DSS_1            M_5            M_4  \\\n",
      "count    4108.000000    4108.000000    4108.000000    4108.000000   \n",
      "mean     9055.580699   20853.144068   34779.817003   21025.322554   \n",
      "std     15644.060978   37979.990394   51855.765387   37119.118065   \n",
      "min         0.000000       0.000000       0.000000       0.000000   \n",
      "25%       219.600000     381.182500     830.930000     777.745000   \n",
      "50%      2274.380000    3893.500000   21480.125000    4747.480000   \n",
      "75%     12961.795000   32564.082500   50793.637500   31204.620000   \n",
      "max    184292.300000  482320.590000  628299.210000  454560.360000   \n",
      "\n",
      "                 M_3            M_2            M_1          Con_4  \\\n",
      "count    4108.000000    4108.000000    4108.000000    4108.000000   \n",
      "mean    15386.404141   20536.753863   17702.385183   16623.431826   \n",
      "std     26093.894008   38278.238001   32111.250898   29175.830089   \n",
      "min         0.000000       0.000000       0.000000       0.000000   \n",
      "25%       760.800000     602.000000     544.420000     242.330000   \n",
      "50%      4126.160000    3074.330000    3132.715000    2798.535000   \n",
      "75%     21892.162500   32388.062500   27433.642500   24359.035000   \n",
      "max    324547.760000  484900.200000  407020.580000  378441.490000   \n",
      "\n",
      "               Con_3          Con_2          Con_1  \n",
      "count    4108.000000    4108.000000    4108.000000  \n",
      "mean    16596.235820   16788.848566   14057.522169  \n",
      "std     29213.555035   30128.574318   25989.113676  \n",
      "min         0.000000       0.000000       0.000000  \n",
      "25%       314.750000     324.375000     217.590000  \n",
      "50%      2813.575000    2453.910000    2272.215000  \n",
      "75%     23853.847500   24971.792500   19098.742500  \n",
      "max    353044.710000  361460.250000  316657.840000  \n",
      "-----------------*END DESCRIBE*-----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "----------*END OVERVIEW OF DATASET 1*-----------\n"
     ]
    }
   ],
   "source": [
    "# Overview of Dataset 1\n",
    "\n",
    "# Get basic information about the dataset\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 1*----------')\n",
    "print('--------------------*INFO*----------------------')\n",
    "print(ds_1.info())\n",
    "print('------------------*END INFO*--------------------')\n",
    "\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# View column names\n",
    "print('-------------------*COLUMNS*--------------------')\n",
    "print(ds_1.columns)\n",
    "print('-----------------*END COLUMNS*------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# List all the columns and their data types\n",
    "print('------------------*DATA TYPES*------------------')\n",
    "print(ds_1.dtypes) \n",
    "print('----------------*END DATA TYPES*----------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Display the first few rows to inspect the data\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_1.head())\n",
    "print('-------------------*END HEAD*-------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Check the summary statistics of numeric columns\n",
    "print('-------------------*DESCRIBE*-------------------')\n",
    "print(ds_1.describe())\n",
    "print('-----------------*END DESCRIBE*-----------------')\n",
    "print('\\n\\n\\n')\n",
    "print('----------*END OVERVIEW OF DATASET 1*-----------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5473977c-acfd-44dc-9602-6862d9c8aa05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---\n",
      "0       COG0001\n",
      "1       COG0002\n",
      "2       COG0003\n",
      "3       COG0004\n",
      "4       COG0005\n",
      "         ...   \n",
      "4103    COG5660\n",
      "4104    COG5661\n",
      "4105    COG5662\n",
      "4106    COG5663\n",
      "4107    COG5664\n",
      "Name: function, Length: 4108, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------\n",
      "Number of rows in the second column: 4108\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the first three columns because they appear of interest from HEAD, starting with column 1\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds1_first_column = ds_1.iloc[:, 0]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds1_row_count_1 = ds1_first_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---')\n",
    "print(ds1_first_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------')\n",
    "print(f\"Number of rows in the second column: {ds1_row_count_1}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9cdecbd6-3e7c-44df-a086-4188b5314d48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 1*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---\n",
      "0               Glutamate-1-semialdehyde aminotransferase\n",
      "1              N-acetyl-gamma-glutamylphosphate reductase\n",
      "2             Anion-transporting ATPase, ArsA/GET3 family\n",
      "3                            Ammonia channel protein AmtB\n",
      "4                         Purine nucleoside phosphorylase\n",
      "                              ...                        \n",
      "4103    Predicted anti-sigma-YlaC factor YlaD, contain...\n",
      "4104             Predicted secreted Zn-dependent protease\n",
      "4105    Transmembrane transcriptional regulator (anti-...\n",
      "4106             Uncharacterized protein, HAD superfamily\n",
      "4107             Predicted secreted Zn-dependent protease\n",
      "Name: description, Length: 4108, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------\n",
      "Number of rows in the second column: 4108\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----\n"
     ]
    }
   ],
   "source": [
    "# Column 2 from HEAD\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 1*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds_1_second_column = ds_1.iloc[:, 1]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds_1_row_count_2 = ds_1_second_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---')\n",
    "print(ds_1_second_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------')\n",
    "print(f\"Number of rows in the second column: {ds_1_row_count_2}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "fcc7db6e-3e53-4142-a700-8078f6ea8af7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 1*----------\n",
      "--*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*--\n",
      "0       28746.42\n",
      "1       47780.21\n",
      "2         580.28\n",
      "3       52589.01\n",
      "4       51522.40\n",
      "          ...   \n",
      "4103       30.73\n",
      "4104        3.50\n",
      "4105       52.84\n",
      "4106      346.50\n",
      "4107        7.00\n",
      "Name: DSS_6, Length: 4108, dtype: float64\n",
      "\n",
      "\n",
      "*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-----*TOTAL NUMBER OF ROWS OF COLUMN THREE*-----\n",
      "Number of rows in the second column: 4108\n",
      "---*END TOTAL NUMBER OF ROWS OF COLUMN THREE*---\n"
     ]
    }
   ],
   "source": [
    "# Column 3 from HEAD\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 1*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds_1_third_column = ds_1.iloc[:, 2]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds_1_row_count_3 = ds_1_third_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('--*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*--')\n",
    "print(ds_1_third_column)\n",
    "print('\\n')\n",
    "print('*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('-----*TOTAL NUMBER OF ROWS OF COLUMN THREE*-----')\n",
    "print(f\"Number of rows in the second column: {ds_1_row_count_3}\")\n",
    "print('---*END TOTAL NUMBER OF ROWS OF COLUMN THREE*---')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb43747-4edf-428f-a9a8-4f64f67d0629",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.2 Dataset One Cleaning</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "654fbbe6-256b-4b13-9a4e-1ce5c6382b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " function       0\n",
      "description    0\n",
      "DSS_6          0\n",
      "DSS_5          0\n",
      "DSS_4          0\n",
      "DSS_3          0\n",
      "DSS_2          0\n",
      "DSS_1          0\n",
      "M_5            0\n",
      "M_4            0\n",
      "M_3            0\n",
      "M_2            0\n",
      "M_1            0\n",
      "Con_4          0\n",
      "Con_3          0\n",
      "Con_2          0\n",
      "Con_1          0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "missing_values1 = ds_1.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e867727e-a9a4-4e8d-8d88-debfbe9a9248",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows: 0\n"
     ]
    }
   ],
   "source": [
    "duplicates_1 = ds_1.duplicated().sum()\n",
    "print(f\"Number of duplicate rows: {duplicates_1}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cecf9fb9-3751-4e76-8324-1c9d0776e3ac",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.2.1 Dataset 1 Key Concepts</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "**Alzheimer's disease:** \"Alzheimer's disease (AD) is a progressive neurodegenerative disorder that is characterized by cognitive decline and the presence of two core pathologies, amyloid β plaques and neurofibrillary tangles. Over the last decade, the presence of a sustained immune response in the brain has emerged as a third core pathology in AD\"<a href=\"#mla_115\"> [115]</a>.\n",
    "\n",
    "**Inflammation:** \"Inflammation is a biological response of the immune system that can be triggered by a variety of factors, including pathogens, damaged cells and toxic compounds. These factors may induce acute and/or chronic inflammatory responses in the heart, pancreas, liver, kidney, lung, brain, intestinal tract and reproductive system, potentially leading to tissue damage or disease\"<a href=\"#mla_116\"> [116]</a>.\n",
    "\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">Sustained overactive inflammation is causal to Alzheimer's, a sustained immune response is indicative of overactive inflammation <a href=\"#mla_117\"> [117]</a><a href=\"#mla_118\"> [118]</a>.</blockquote><br>\n",
    "\n",
    "**Oxidative Stress:** The phenomena of Alzheimer's \"are mainly initiated and enhanced by oxidative stress, a process referring to an imbalance between antioxidants and oxidants in favour of oxidants. This imbalance can occur as a result of increased free radicals or a decrease in antioxidant defense, free radicals being a species that contains one or more unpaired electrons in its outer shell\"<a href=\"#mla_119\"> [119]</a>.\n",
    "\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">An imbalanced oxidative stress response is causal to Alzheimer's, such as having too much systemic oxidative stress<a href=\"#mla_120\"> [120]</a>.</blockquote><br>\n",
    "\n",
    "**The Gut Microbiome & Inflammation:** \"The gut microbiota encompasses a diverse community of bacteria that carry out various functions influencing the overall health of the host. These comprise nutrient metabolism, immune system regulation and natural defense against infection. The presence of certain bacteria is associated with inflammatory molecules that may bring about inflammation in various body tissues. Inflammation underlies many chronic multisystem conditions including obesity, atherosclerosis, type 2 diabetes mellitus and inflammatory bowel disease. Inflammation may be triggered by structural components of the bacteria which can result in a cascade of inflammatory pathways involving interleukins and other cytokines. Similarly, by-products of metabolic processes in bacteria, including some short-chain fatty acids, can play a role in inhibiting inflammatory processes\"<a href=\"#mla_121\"> [121]</a>.\n",
    "\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">A well balanced good gut microbiome is anti-inflammatory, which prevents diseases such as Alzheimer's<a href=\"#mla_122\"> [122]</a>.</blockquote><br>\n",
    "\n",
    "**The Gut Microbiome & Oxidative Stress:** \"Commensal (good) bacteria often exhibit anti-oxidative properties and suppress inflammatory reactions. Pathogenic microbiota induce inflammation and shift the ... balance toward a pro-oxidative status\"<a href=\"#mla_123\"> [123]</a>.\n",
    "\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">Having more good gut bacteria is not only anti-inflammatory, it is anti-oxidative, having more \"bad\" gut bacteria is pro-oxidative<a href=\"#mla_124\"> [124]</a>.</blockquote><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51c24fd9-03ad-413e-b78f-77c08b69e8f0",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.2.2 Dataset 1 Abstract</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "This dataset is taken from a study that aims to see the effects of a Traditional Chinese Medicine Herb known as Danggui Shaoyao San, on modulating and even preventing Alzheimer's Disease.\n",
    "\n",
    "Danggui Shaoyao San has been used for centuries in Traditional Chinese Medicine, for a variety of purposes, and is composed of 6 herbs. \n",
    "Angelica sinensis (Oliv.) Diels (Umbelliferae), Paeonia lactiflora Pall. (Paeoniaceae), Ligusticum chuanxiong Hort. (Umbelliferae), Poria cocos (Schw.) Wolf (Polyporaceae), Atractylodes macrocephala Koidz. (Compositae), and Alisma orientalis (Sam.) Juzep. (Alismataceae)<a href=\"#mla_125\"> [125]</a>.\n",
    "\n",
    "<h4>Evidence-based Health Benefits of Danggui Shaoyao San:</h4>\n",
    "\n",
    "- 1.) Cognitive Function and Dementia: Has neuroprotective effects, which reduces the risk of dementia<a href=\"#mla_126\"> [126]</a>.\n",
    "- 2.) Digestive Health: Promotes healthy digestion and stomach function<a href=\"#mla_129\"> [129]</a>.\n",
    "- 3.) Anti-inflammatory and Antioxidant Effects: Reduces inflammation and oxidative stress in the body<a href=\"#mla_127\"> [127]</a>.\n",
    "- 4.) Stress and Anxiety Relief: Paeonia lactiflora, may help reduce stress and anxiety, promoting mental relaxation<a href=\"#mla_128\"> [128]</a>.\n",
    "\n",
    "This study was performed on 30 Thirty Sprague-Dawley (SD) rats, that where separated into three main groups. Two rat groups were injected via bilateral intracerebroventricular injections of streptozotocin (STZ). STZ has toxic effects on insulin-producing cells, which induces insulin resistance for the rats. With induced insulin resistance, and thus impaired brain-insulin sensitivity, Alzheimer's is induced. Alzheimer's is induced in two groups, and there is one control group<a href=\"#mla_130\"> [130]</a>. \n",
    "\n",
    "\"Bloodstream metabolites were characterized, gut microbiota profiled through 16S rDNA sequencing, and cortical metabolomics analyzed. Hippocampal proinflammatory cytokines (IL-1β, IL-6, TNF-α) were quantified using RT-qPCR, and oxidative stress markers (SOD, CAT, GSH-PX, MDA) in brain tissues were measured with biochemical assays\"<a href=\"#mla_131\"> [131]</a>.\n",
    "\n",
    "\"The top 45 most abundant species in the gut microbiota were determined across these three groups. Subsequently, Spearman correlation analysis and a heatmap were utilized to assess the correlation between gut microbiota composition and metabolite levels\"<a href=\"#mla_132\"> [132]</a>.\n",
    "\n",
    "\n",
    "<h4>Test Groups for Dataset 1 (columns):</h4>  \n",
    "\n",
    "- 1.) DSS: A rat group that receives Danggui Shaoyao San (DSS), and is induced with Alzheimer's.\n",
    "- 2.) M: A rat group that receives zero Danggui Shaoyao San (DSS), and is induced with Alzheimer's.\n",
    "- 3.) Con: A control rat group that receives zero Danggui Shaoyao San (DSS), and is not induced with Alzheimer's.\n",
    "- 4.) Description: Biomarkers\n",
    "\n",
    "<h4>Conclusion</h4>\n",
    "\n",
    "Thousands of markers look at bloodstream metabolites, microbiota metabolites, inflammatory cytokine markers, and antioxidative enzyme markers, to draw inferences from how the gut-brain axis affects oxidative stress and neuroinflammation, both of which have been shown to cause Alzheimer's disease. The depth of informative biomarkers from this dataset is quite vast."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe9f15a7-ea14-4205-b663-4d4b4dbcb8bc",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.3 Dataset Two</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53d7b83a-c4dc-44bd-afe7-63b87c3964a4",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.4 Dataset Two Overview</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6ea674c1-4e56-4760-aac6-865fa75585da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "--------------------*INFO*----------------------\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 131 entries, 0 to 130\n",
      "Data columns (total 23 columns):\n",
      " #   Column                                                                                          Non-Null Count  Dtype \n",
      "---  ------                                                                                          --------------  ----- \n",
      " 0   Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  130 non-null    object\n",
      " 1   Unnamed: 1                                                                                      130 non-null    object\n",
      " 2   Unnamed: 2                                                                                      130 non-null    object\n",
      " 3   Unnamed: 3                                                                                      130 non-null    object\n",
      " 4   Unnamed: 4                                                                                      130 non-null    object\n",
      " 5   Unnamed: 5                                                                                      130 non-null    object\n",
      " 6   Unnamed: 6                                                                                      130 non-null    object\n",
      " 7   Unnamed: 7                                                                                      130 non-null    object\n",
      " 8   Unnamed: 8                                                                                      130 non-null    object\n",
      " 9   Unnamed: 9                                                                                      130 non-null    object\n",
      " 10  Unnamed: 10                                                                                     130 non-null    object\n",
      " 11  Unnamed: 11                                                                                     130 non-null    object\n",
      " 12  Unnamed: 12                                                                                     130 non-null    object\n",
      " 13  Unnamed: 13                                                                                     130 non-null    object\n",
      " 14  Unnamed: 14                                                                                     130 non-null    object\n",
      " 15  Unnamed: 15                                                                                     130 non-null    object\n",
      " 16  Unnamed: 16                                                                                     130 non-null    object\n",
      " 17  Unnamed: 17                                                                                     130 non-null    object\n",
      " 18  Unnamed: 18                                                                                     130 non-null    object\n",
      " 19  Unnamed: 19                                                                                     130 non-null    object\n",
      " 20  Unnamed: 20                                                                                     130 non-null    object\n",
      " 21  Unnamed: 21                                                                                     130 non-null    object\n",
      " 22  Unnamed: 22                                                                                     130 non-null    object\n",
      "dtypes: object(23)\n",
      "memory usage: 23.7+ KB\n",
      "None\n",
      "------------------*END INFO*--------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*COLUMNS*--------------------\n",
      "Index(['Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.',\n",
      "       'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5',\n",
      "       'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10',\n",
      "       'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14',\n",
      "       'Unnamed: 15', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',\n",
      "       'Unnamed: 19', 'Unnamed: 20', 'Unnamed: 21', 'Unnamed: 22'],\n",
      "      dtype='object')\n",
      "-----------------*END COLUMNS*------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------------------*DATA TYPES*------------------\n",
      "Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.    object\n",
      "Unnamed: 1                                                                                        object\n",
      "Unnamed: 2                                                                                        object\n",
      "Unnamed: 3                                                                                        object\n",
      "Unnamed: 4                                                                                        object\n",
      "Unnamed: 5                                                                                        object\n",
      "Unnamed: 6                                                                                        object\n",
      "Unnamed: 7                                                                                        object\n",
      "Unnamed: 8                                                                                        object\n",
      "Unnamed: 9                                                                                        object\n",
      "Unnamed: 10                                                                                       object\n",
      "Unnamed: 11                                                                                       object\n",
      "Unnamed: 12                                                                                       object\n",
      "Unnamed: 13                                                                                       object\n",
      "Unnamed: 14                                                                                       object\n",
      "Unnamed: 15                                                                                       object\n",
      "Unnamed: 16                                                                                       object\n",
      "Unnamed: 17                                                                                       object\n",
      "Unnamed: 18                                                                                       object\n",
      "Unnamed: 19                                                                                       object\n",
      "Unnamed: 20                                                                                       object\n",
      "Unnamed: 21                                                                                       object\n",
      "Unnamed: 22                                                                                       object\n",
      "dtype: object\n",
      "----------------*END DATA TYPES*----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "---------------------*HEAD*---------------------\n",
      "  Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "0                                                NaN                                               \n",
      "1                                            Samples                                               \n",
      "2                                          GSM300167                                               \n",
      "3                                          GSM300168                                               \n",
      "4                                          GSM300169                                               \n",
      "\n",
      "               Unnamed: 1 Unnamed: 2   Unnamed: 3 Unnamed: 4 Unnamed: 5  \\\n",
      "0                     NaN        NaN          NaN        NaN        NaN   \n",
      "1            Brain region     Gender  Braak stage      ATP5B       BDNF   \n",
      "2  superior frontal gyrus     female            0   0.671874    0.80707   \n",
      "3             hippocampus     female            0   0.972526   2.671879   \n",
      "4             hippocampus       male            0   0.744222    3.33928   \n",
      "\n",
      "  Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14  \\\n",
      "0        NaN        NaN        NaN        NaN  ...         NaN         NaN   \n",
      "1       CD44        DLD     GABRG2       GAD2  ...      PSMD14     SLC32A1   \n",
      "2   1.981505   0.732005    0.32877   0.641998  ...    0.681275    0.165948   \n",
      "3   7.455322   1.066505   0.640967   0.352101  ...    1.247155    1.211095   \n",
      "4   3.321804   0.877786   0.357565   0.317235  ...    1.120629    0.621519   \n",
      "\n",
      "  Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \\\n",
      "0         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "1       SMAD4      SNAP25         SYP        SYT1        TJP1       VDAC1   \n",
      "2    0.857919    0.914064    0.643572    0.685069    1.045188    0.557694   \n",
      "3    0.734436    0.752652    0.912192    0.710386    0.765581    0.845077   \n",
      "4     0.96861    0.690655    0.757741    0.564051    0.904162    0.558936   \n",
      "\n",
      "  Unnamed: 21 Unnamed: 22  \n",
      "0         NaN         NaN  \n",
      "1       WWTR1        YAP1  \n",
      "2    1.467627    1.755059  \n",
      "3    1.629639    1.158677  \n",
      "4    1.518328    1.704543  \n",
      "\n",
      "[5 rows x 23 columns]\n",
      "-------------------*END HEAD*-------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*DESCRIBE*-------------------\n",
      "       Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "count                                                 130                                               \n",
      "unique                                                130                                               \n",
      "top                                               Samples                                               \n",
      "freq                                                    1                                               \n",
      "\n",
      "                    Unnamed: 1 Unnamed: 2  Unnamed: 3 Unnamed: 4 Unnamed: 5  \\\n",
      "count                      130        130         130        130        130   \n",
      "unique                       3          3           5        130        130   \n",
      "top     superior frontal gyrus     female           0      ATP5B       BDNF   \n",
      "freq                        68         67          91          1          1   \n",
      "\n",
      "       Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ...  Unnamed: 13  \\\n",
      "count         130        130        130        130  ...          130   \n",
      "unique        130        130        130        130  ...          130   \n",
      "top          CD44        DLD     GABRG2       GAD2  ...       PSMD14   \n",
      "freq            1          1          1          1  ...            1   \n",
      "\n",
      "       Unnamed: 14 Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18  \\\n",
      "count          130         130         130         130         130   \n",
      "unique         130         130         130         130         130   \n",
      "top        SLC32A1       SMAD4      SNAP25         SYP        SYT1   \n",
      "freq             1           1           1           1           1   \n",
      "\n",
      "       Unnamed: 19 Unnamed: 20 Unnamed: 21 Unnamed: 22  \n",
      "count          130         130         130         130  \n",
      "unique         130         130         130         130  \n",
      "top           TJP1       VDAC1       WWTR1        YAP1  \n",
      "freq             1           1           1           1  \n",
      "\n",
      "[4 rows x 23 columns]\n",
      "-----------------*END DESCRIBE*-----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "----------*END OVERVIEW OF DATASET 2*-----------\n"
     ]
    }
   ],
   "source": [
    "# Overview of Dataset 1\n",
    "\n",
    "# Get basic information about the dataset\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "print('--------------------*INFO*----------------------')\n",
    "print(ds_2.info())\n",
    "print('------------------*END INFO*--------------------')\n",
    "\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# View column names\n",
    "print('-------------------*COLUMNS*--------------------')\n",
    "print(ds_2.columns)\n",
    "print('-----------------*END COLUMNS*------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# List all the columns and their data types\n",
    "print('------------------*DATA TYPES*------------------')\n",
    "print(ds_2.dtypes) \n",
    "print('----------------*END DATA TYPES*----------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Display the first few rows to inspect the data\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_2.head())\n",
    "print('-------------------*END HEAD*-------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Check the summary statistics of numeric columns\n",
    "print('-------------------*DESCRIBE*-------------------')\n",
    "print(ds_2.describe())\n",
    "print('-----------------*END DESCRIBE*-----------------')\n",
    "print('\\n\\n\\n')\n",
    "print('----------*END OVERVIEW OF DATASET 2*-----------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "2eb032f8-941a-4c69-b799-532ff4139cc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---\n",
      "0             NaN\n",
      "1         Samples\n",
      "2       GSM300167\n",
      "3       GSM300168\n",
      "4       GSM300169\n",
      "          ...    \n",
      "126    GSM1176262\n",
      "127    GSM1176264\n",
      "128    GSM1176266\n",
      "129    GSM1176269\n",
      "130    GSM1176274\n",
      "Name: Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF., Length: 131, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------\n",
      "Number of rows in the second column: 130\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the first four columns because they appear of interest from HEAD, starting with column 1\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds2_first_column = ds_2.iloc[:, 0]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds2_row_count_1 = ds2_first_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---')\n",
    "print(ds2_first_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------')\n",
    "print(f\"Number of rows in the second column: {ds2_row_count_1}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1f8919d4-62d5-4514-849e-ad3ba6b6222a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---\n",
      "0                         NaN\n",
      "1                Brain region\n",
      "2      superior frontal gyrus\n",
      "3                 hippocampus\n",
      "4                 hippocampus\n",
      "                ...          \n",
      "126    superior frontal gyrus\n",
      "127    superior frontal gyrus\n",
      "128    superior frontal gyrus\n",
      "129    superior frontal gyrus\n",
      "130    superior frontal gyrus\n",
      "Name: Unnamed: 1, Length: 131, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------\n",
      "Number of rows in the second column: 130\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the second column because it appears of interest from HEAD\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds2_second_column = ds_2.iloc[:, 1]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds2_row_count_2 = ds2_second_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---')\n",
    "print(ds2_second_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------')\n",
    "print(f\"Number of rows in the second column: {ds2_row_count_2}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c204af36-a22b-4a75-a1d2-3f9845e204e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "--*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*--\n",
      "0         NaN\n",
      "1      Gender\n",
      "2      female\n",
      "3      female\n",
      "4        male\n",
      "        ...  \n",
      "126    female\n",
      "127    female\n",
      "128    female\n",
      "129      male\n",
      "130      male\n",
      "Name: Unnamed: 2, Length: 131, dtype: object\n",
      "\n",
      "\n",
      "*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-----*TOTAL NUMBER OF ROWS OF COLUMN THREE*-----\n",
      "Number of rows in the second column: 130\n",
      "---*END TOTAL NUMBER OF ROWS OF COLUMN THREE*---\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the third column because it appears of interest from HEAD\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds2_third_column = ds_2.iloc[:, 2]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds2_row_count_3 = ds2_third_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('--*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*--')\n",
    "print(ds2_third_column)\n",
    "print('\\n')\n",
    "print('*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('-----*TOTAL NUMBER OF ROWS OF COLUMN THREE*-----')\n",
    "print(f\"Number of rows in the second column: {ds2_row_count_3}\")\n",
    "print('---*END TOTAL NUMBER OF ROWS OF COLUMN THREE*---')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "227c3d39-9bcc-494b-97b2-1798a31f3e8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 2*----------\n",
      "--*INSPECTION OF NUMBER OF ROWS : COLUMN FOUR*--\n",
      "0              NaN\n",
      "1      Braak stage\n",
      "2                0\n",
      "3                0\n",
      "4                0\n",
      "          ...     \n",
      "126           V-VI\n",
      "127           V-VI\n",
      "128           V-VI\n",
      "129           V-VI\n",
      "130           V-VI\n",
      "Name: Unnamed: 3, Length: 131, dtype: object\n",
      "\n",
      "\n",
      "*END INSPECTION OF NUMBER OF ROWS : COLUMN FOUR*\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN FOUR*-----\n",
      "Number of rows in the second column: 130\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN FOUR*---\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the forth column because it appears of interest from HEAD\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 2*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds2_fourth_column = ds_2.iloc[:, 3]  # 0-based index, so 1 is the second column\n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds2_row_count_4 = ds2_fourth_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('--*INSPECTION OF NUMBER OF ROWS : COLUMN FOUR*--')\n",
    "print(ds2_fourth_column)\n",
    "print('\\n')\n",
    "print('*END INSPECTION OF NUMBER OF ROWS : COLUMN FOUR*')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN FOUR*-----')\n",
    "print(f\"Number of rows in the second column: {ds2_row_count_4}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN FOUR*---')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b69f55ab-0a50-4fd7-9759-37d9b93961ef",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.2 Dataset Two Cleaning</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d421e464-2a39-4b71-a31f-9028457debe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.    1\n",
      "Unnamed: 1                                                                                        1\n",
      "Unnamed: 2                                                                                        1\n",
      "Unnamed: 3                                                                                        1\n",
      "Unnamed: 4                                                                                        1\n",
      "Unnamed: 5                                                                                        1\n",
      "Unnamed: 6                                                                                        1\n",
      "Unnamed: 7                                                                                        1\n",
      "Unnamed: 8                                                                                        1\n",
      "Unnamed: 9                                                                                        1\n",
      "Unnamed: 10                                                                                       1\n",
      "Unnamed: 11                                                                                       1\n",
      "Unnamed: 12                                                                                       1\n",
      "Unnamed: 13                                                                                       1\n",
      "Unnamed: 14                                                                                       1\n",
      "Unnamed: 15                                                                                       1\n",
      "Unnamed: 16                                                                                       1\n",
      "Unnamed: 17                                                                                       1\n",
      "Unnamed: 18                                                                                       1\n",
      "Unnamed: 19                                                                                       1\n",
      "Unnamed: 20                                                                                       1\n",
      "Unnamed: 21                                                                                       1\n",
      "Unnamed: 22                                                                                       1\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "missing_values_2 = ds_2.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7b73d793-face-417f-a5c7-e303b6166535",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows with any null values:\n",
      "   Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "0                                                NaN                                               \n",
      "\n",
      "  Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \\\n",
      "0        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "\n",
      "  Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14 Unnamed: 15  \\\n",
      "0        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "\n",
      "  Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20 Unnamed: 21  \\\n",
      "0         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "\n",
      "  Unnamed: 22  \n",
      "0         NaN  \n",
      "\n",
      "[1 rows x 23 columns]\n",
      "Rows with null values in 'column_name':\n",
      "   Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "0                                                NaN                                               \n",
      "\n",
      "  Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \\\n",
      "0        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "\n",
      "  Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14 Unnamed: 15  \\\n",
      "0        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "\n",
      "  Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20 Unnamed: 21  \\\n",
      "0         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "\n",
      "  Unnamed: 22  \n",
      "0         NaN  \n",
      "\n",
      "[1 rows x 23 columns]\n",
      "Exact data points that are null:\n",
      "     Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "0                                                  NaN                                               \n",
      "1                                                  NaN                                               \n",
      "2                                                  NaN                                               \n",
      "3                                                  NaN                                               \n",
      "4                                                  NaN                                               \n",
      "..                                                 ...                                               \n",
      "126                                                NaN                                               \n",
      "127                                                NaN                                               \n",
      "128                                                NaN                                               \n",
      "129                                                NaN                                               \n",
      "130                                                NaN                                               \n",
      "\n",
      "    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \\\n",
      "0          NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "1          NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "2          NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "3          NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "4          NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "..         ...        ...        ...        ...        ...        ...   \n",
      "126        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "127        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "128        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "129        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "130        NaN        NaN        NaN        NaN        NaN        NaN   \n",
      "\n",
      "    Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14 Unnamed: 15  \\\n",
      "0          NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "1          NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "2          NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "3          NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "4          NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "..         ...        ...        ...  ...         ...         ...         ...   \n",
      "126        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "127        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "128        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "129        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "130        NaN        NaN        NaN  ...         NaN         NaN         NaN   \n",
      "\n",
      "    Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20 Unnamed: 21  \\\n",
      "0           NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "1           NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "2           NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "3           NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "4           NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "..          ...         ...         ...         ...         ...         ...   \n",
      "126         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "127         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "128         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "129         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "130         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "\n",
      "    Unnamed: 22  \n",
      "0           NaN  \n",
      "1           NaN  \n",
      "2           NaN  \n",
      "3           NaN  \n",
      "4           NaN  \n",
      "..          ...  \n",
      "126         NaN  \n",
      "127         NaN  \n",
      "128         NaN  \n",
      "129         NaN  \n",
      "130         NaN  \n",
      "\n",
      "[131 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "rows_with_nulls_2 = ds_2[ds_2.isnull().any(axis=1)]\n",
    "print(\"Rows with any null values:\\n\", rows_with_nulls_2)\n",
    "\n",
    "# View rows where a specific column has null values (e.g., 'column_name')\n",
    "rows_with_nulls_in_specific_column_2 = ds_2[ds_2['Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.'].isnull()]\n",
    "print(f\"Rows with null values in 'column_name':\\n\", rows_with_nulls_in_specific_column_2)\n",
    "\n",
    "# View the exact data points that are null\n",
    "null_data_points_2 = ds_2[ds_2.isnull()]\n",
    "print(\"Exact data points that are null:\\n\", null_data_points_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c93a2830-2b3f-4007-8f0d-be8ae838626f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.  \\\n",
      "1                                            Samples                                               \n",
      "2                                          GSM300167                                               \n",
      "3                                          GSM300168                                               \n",
      "4                                          GSM300169                                               \n",
      "5                                          GSM300170                                               \n",
      "\n",
      "               Unnamed: 1 Unnamed: 2   Unnamed: 3 Unnamed: 4 Unnamed: 5  \\\n",
      "1            Brain region     Gender  Braak stage      ATP5B       BDNF   \n",
      "2  superior frontal gyrus     female            0   0.671874    0.80707   \n",
      "3             hippocampus     female            0   0.972526   2.671879   \n",
      "4             hippocampus       male            0   0.744222    3.33928   \n",
      "5             hippocampus       male            0   0.890069   3.934139   \n",
      "\n",
      "  Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 13 Unnamed: 14  \\\n",
      "1       CD44        DLD     GABRG2       GAD2  ...      PSMD14     SLC32A1   \n",
      "2   1.981505   0.732005    0.32877   0.641998  ...    0.681275    0.165948   \n",
      "3   7.455322   1.066505   0.640967   0.352101  ...    1.247155    1.211095   \n",
      "4   3.321804   0.877786   0.357565   0.317235  ...    1.120629    0.621519   \n",
      "5    7.20695    0.77125   0.342878   0.312362  ...    1.264295    0.610008   \n",
      "\n",
      "  Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \\\n",
      "1       SMAD4      SNAP25         SYP        SYT1        TJP1       VDAC1   \n",
      "2    0.857919    0.914064    0.643572    0.685069    1.045188    0.557694   \n",
      "3    0.734436    0.752652    0.912192    0.710386    0.765581    0.845077   \n",
      "4     0.96861    0.690655    0.757741    0.564051    0.904162    0.558936   \n",
      "5    0.852232    0.667956    0.823295     0.60283    0.657229    0.605209   \n",
      "\n",
      "  Unnamed: 21 Unnamed: 22  \n",
      "1       WWTR1        YAP1  \n",
      "2    1.467627    1.755059  \n",
      "3    1.629639    1.158677  \n",
      "4    1.518328    1.704543  \n",
      "5    1.765773    1.283766  \n",
      "\n",
      "[5 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "ds_2 = ds_2.drop(index=0)\n",
    "\n",
    "# Display the dataset after dropping row 0\n",
    "print(ds_2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e0b0a03d-89db-4ff9-9bfd-5b5d77958a00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " Supplementary table 4: Hub genes expression in different Braak stage of AD in the HPC and SPF.    0\n",
      "Unnamed: 1                                                                                        0\n",
      "Unnamed: 2                                                                                        0\n",
      "Unnamed: 3                                                                                        0\n",
      "Unnamed: 4                                                                                        0\n",
      "Unnamed: 5                                                                                        0\n",
      "Unnamed: 6                                                                                        0\n",
      "Unnamed: 7                                                                                        0\n",
      "Unnamed: 8                                                                                        0\n",
      "Unnamed: 9                                                                                        0\n",
      "Unnamed: 10                                                                                       0\n",
      "Unnamed: 11                                                                                       0\n",
      "Unnamed: 12                                                                                       0\n",
      "Unnamed: 13                                                                                       0\n",
      "Unnamed: 14                                                                                       0\n",
      "Unnamed: 15                                                                                       0\n",
      "Unnamed: 16                                                                                       0\n",
      "Unnamed: 17                                                                                       0\n",
      "Unnamed: 18                                                                                       0\n",
      "Unnamed: 19                                                                                       0\n",
      "Unnamed: 20                                                                                       0\n",
      "Unnamed: 21                                                                                       0\n",
      "Unnamed: 22                                                                                       0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "missing_values_2 = ds_2.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b95cbfd5-9547-4adb-a0d0-88f175b806d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows: 0\n"
     ]
    }
   ],
   "source": [
    "# Check for duplicates\n",
    "duplicates_2 = ds_2.duplicated().sum()\n",
    "print(f\"Number of duplicate rows: {duplicates_2}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff45d61e-100e-4b46-86f2-f2f55d1d08cf",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.4.1 Dataset 2 Key Concepts</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "**Differentially expressed genes (DEGs):** Refer to the process used in bioinformatics and genomics to discover genes that show significant differences in expression levels between different conditions, groups, or treatments. For example, DEGs could be genes that are upregulated (increased expression) or downregulated (decreased expression) in cancerous tissue compared to normal tissue<a href=\"#mla_133\"> [133]</a>.\n",
    "\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">Disease and health both have differentially expressed genes, especially in Alzheimer's<a href=\"#mla_134\"> [134]</a>.</blockquote><br>\n",
    "\n",
    "\n",
    "**Microarray Experiment:** A microarray experiment uses technology to measure the expression levels of many genes simultaneously. There are thousands of microscopic spots in a microchip array, that are filled with a specific gene sequence that is being searched for expression levels<a href=\"#mla_135\"> [135]</a>.\n",
    "\n",
    "**Gene Ontology:** is a comprehensive and standardized system that categorizes gene functions into biological processes, molecular functions, and cellular components. It allows researchers to annotate genes consistently, compare gene functions across species, and analyze biological data in a systematic way, leading to a deeper understanding of gene roles in various biological contexts<a href=\"#mla_136\"> [136]</a>.\n",
    "\n",
    "**Protein-Protein Interaction (PPI) Network:** are a graphical representation of the physical and functional interactions between proteins within a cell or organism<a href=\"#mla_137\"> [137]</a>.\n",
    "\n",
    "**Hub genes:** are genes that play central roles in biological networks, often characterized by having a large number of connections (high degree) within a network<a href=\"#mla_138\"> [138]</a>.\n",
    "\n",
    "**Brain-derived neurotrophic Factor (BDNF):** is a protein that plays a critical role in neuronal survival, neurogenesis, synaptic plasticity, and cognitive functions. Without BDNF neurodegenerative diseases like Alzheimer's disease, Parkinson's disease, and Huntington's disease may contribute to the progression of these conditions by impairing neuronal survival and function. It is one of the most important neurotrophins, a family of proteins that promote the survival, development, and function of neurons<a href=\"#mla_139\"> [139]</a>.\n",
    "\n",
    "**WW domain-containing transcription regulator protein 1 (WWTR1):** is a protein domain found in many proteins involved in cellular signaling and regulation. It mediates protein-protein interactions and plays a role in cellular processes such as cell growth, apoptosis, and differentiation. It is key to keep appropriate organ size, and plays a role in cancer<a href=\"#mla_140\"> [140]</a>.\n",
    "\n",
    "**Braak stages:** refer to a system used to describe the progression of Alzheimer's disease pathology in the brain, specifically focusing on the distribution of neurofibrillary tangles, which are abnormal aggregates of tau protein. This staging system helps in understanding the progression of the disease and correlating it with clinical symptoms and cognitive decline<a href=\"#mla_141\"> [141]</a>.\n",
    "\n",
    "- Braak Stage I: Early stage with minimal cognitive impairment, often asymptomatic.\n",
    "- Braak Stage II: Mild cognitive impairment may begin to be noticeable, particularly in areas related to memory.\n",
    "- Braak Stage III: Progressive cognitive decline is evident, with increased memory problems and other symptoms affecting daily life.\n",
    "- Braak Stage IV: More pronounced cognitive deficits, including difficulties with language, spatial orientation, and executive functions.\n",
    "- Braak Stage V: Severe cognitive impairment, with significant loss of independence and function. More extensive involvement of various cognitive domains.\n",
    "- Braak Stage VI: Advanced stage of Alzheimer's disease with profound cognitive impairment and severe functional decline. Patients may experience significant difficulties with basic activities of daily living.\n",
    "\n",
    "\n",
    "**Amyloid-beta (Aβ) 42 & β-secretase activity:** is a specific form of amyloid-beta peptide that is 42 amino acids long. It is one of the most studied peptides in Alzheimer's disease research. Aβ 42 is prone to aggregation and forms amyloid plaques, which are a hallmark of Alzheimer's disease. These plaques are toxic to neurons and are associated with cognitive decline. Elevated levels of Aβ 42 in cerebrospinal fluid (CSF) are often used as a biomarker for Alzheimer's disease. Decreased levels of Aβ 42 in CSF can indicate its increased deposition in the brain, which is characteristic of the disease. β-Secretase activity leads to the production of Aβ peptides, including Aβ 42. Increased β-secretase activity results in higher levels of Aβ peptides, which can contribute to the formation of amyloid plaques in the brain<a href=\"#mla_142\"> [142]</a><a href=\"#mla_143\"> [143]</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b404dc6-45e6-48fb-a560-57d6e3765c37",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.4.2 Dataset 2 Abstract</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "This dataset is a meta-analysis study that analyzes multiple genetic datasets. The study aims to find common differentially expressed genes (DEGs) across multiple microarray experiments (and their resultant datasets), to identify genetic biomarkers, and hub genes for the diagnosis and treatment of Alzheimer's disease (AD).\n",
    "\n",
    "DEGs from five microarray datasets were identified and intersected to \"detect consensus DEGs\"<a href=\"#mla_144\"> [144]</a>. Based on these consensus DEGs a gene Ontology was conducted, and hub genes where identified with the open source platform Cytoscape for visualizing complex networks between these genes.\n",
    "\n",
    "There were \"608 consensus DEGs\" and \"18 hub genes\"<a href=\"#mla_145\"> [145]</a>. Biomarkers such as \"Brain-derived neurotrophic factor (BDNF) and WW domain-containing transcription regulator protein 1 (WWTR1), which are associated with the Braak stage, Aβ 42 levels, and β-secretase activity, were identified as critical genes of AD\"<a href=\"#mla_146\"> [146]</a>. The \"study identified 16 hub genes correlated to the neuropathological stage and 35 potential biomarkers for the diagnosis of AD\"<a href=\"#mla_147\"> [147]</a>.\n",
    "\n",
    "<h4>Dataset 1 (columns):</h4>  \n",
    "\n",
    "- 1.) GSMs: The first column lists the mice specimen.\n",
    "- 2.) Brain Part Analyzed: The second column lists the brain part analyzed.\n",
    "- 3.) Gender: Lists the gender of the mice.\n",
    "- 4.) Braak Stage: Lists the Braak Stage that the mice were at at the time of analysis.\n",
    "- 4.) All following columns: Biomarkers\n",
    "\n",
    "<h4>Conclusion</h4>\n",
    "\n",
    "\n",
    "The first four columns are categorical descriptions of specimen information, and then there are 19 columns that describe biomarkers that where under analysis that relate to genetics. These biomarkers where to draw pathological inferences on central genes or \"hub genes\", and genetic pathways that ultimately contribute to the phenomenon of Alzheimer's Disease."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5714b4f3-5f9d-4616-9847-ee13e44de6fd",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.5 Dataset Three</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b4912c-614a-435f-87df-df134fa30c60",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.6 Dataset Three Overview</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "337e7240-2495-4626-874c-61929c12c88a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "--------------------*INFO*----------------------\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 29 entries, 0 to 28\n",
      "Data columns (total 6 columns):\n",
      " #   Column          Non-Null Count  Dtype  \n",
      "---  ------          --------------  -----  \n",
      " 0   Mouse           29 non-null     object \n",
      " 1   Treatment       29 non-null     object \n",
      " 2   Sex             29 non-null     object \n",
      " 3   GADPH Ct Liver  29 non-null     object \n",
      " 4   CRP Liver Ct    12 non-null     float64\n",
      " 5   N/A             0 non-null      float64\n",
      "dtypes: float64(2), object(4)\n",
      "memory usage: 1.5+ KB\n",
      "None\n",
      "------------------*END INFO*--------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*COLUMNS*--------------------\n",
      "Index(['Mouse', 'Treatment', 'Sex', 'GADPH Ct Liver', 'CRP Liver Ct', 'N/A'], dtype='object')\n",
      "-----------------*END COLUMNS*------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------------------*DATA TYPES*------------------\n",
      "Mouse              object\n",
      "Treatment          object\n",
      "Sex                object\n",
      "GADPH Ct Liver     object\n",
      "CRP Liver Ct      float64\n",
      "N/A               float64\n",
      "dtype: object\n",
      "----------------*END DATA TYPES*----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "---------------------*HEAD*---------------------\n",
      "  Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A\n",
      "0    83         W   F          34.75           NaN  NaN\n",
      "1    85         W   F          25.57     29.130000  NaN\n",
      "2    86         W   M          33.91     37.609234  NaN\n",
      "3    87         W   M      37.272731     38.420000  NaN\n",
      "4    89         W   M          28.93     29.660000  NaN\n",
      "-------------------*END HEAD*-------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*DESCRIBE*-------------------\n",
      "       CRP Liver Ct  N/A\n",
      "count     12.000000  0.0\n",
      "mean      32.679405  NaN\n",
      "std        4.249010  NaN\n",
      "min       26.590000  NaN\n",
      "25%       29.527500  NaN\n",
      "50%       31.650479  NaN\n",
      "75%       36.987308  NaN\n",
      "max       38.420000  NaN\n",
      "-----------------*END DESCRIBE*-----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "----------*END OVERVIEW OF DATASET 3*-----------\n"
     ]
    }
   ],
   "source": [
    "# Overview of Dataset 1\n",
    "\n",
    "# Get basic information about the dataset\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "print('--------------------*INFO*----------------------')\n",
    "print(ds_3.info())\n",
    "print('------------------*END INFO*--------------------')\n",
    "\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# View column names\n",
    "print('-------------------*COLUMNS*--------------------')\n",
    "print(ds_3.columns)\n",
    "print('-----------------*END COLUMNS*------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# List all the columns and their data types\n",
    "print('------------------*DATA TYPES*------------------')\n",
    "print(ds_3.dtypes) \n",
    "print('----------------*END DATA TYPES*----------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Display the first few rows to inspect the data\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_3.head())\n",
    "print('-------------------*END HEAD*-------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Check the summary statistics of numeric columns\n",
    "print('-------------------*DESCRIBE*-------------------')\n",
    "print(ds_3.describe())\n",
    "print('-----------------*END DESCRIBE*-----------------')\n",
    "print('\\n\\n\\n')\n",
    "print('----------*END OVERVIEW OF DATASET 3*-----------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a77b2fed-20ba-4e11-aee0-7a2705b6a56b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---\n",
      "0        83\n",
      "1        85\n",
      "2        86\n",
      "3        87\n",
      "4        89\n",
      "5        90\n",
      "6        94\n",
      "7        99\n",
      "8       249\n",
      "9        81\n",
      "10       82\n",
      "11       92\n",
      "12      296\n",
      "13      298\n",
      "14    Mouse\n",
      "15       83\n",
      "16       85\n",
      "17       86\n",
      "18       87\n",
      "19       89\n",
      "20       90\n",
      "21       94\n",
      "22       99\n",
      "23      249\n",
      "24       81\n",
      "25       82\n",
      "26       92\n",
      "27      296\n",
      "28      298\n",
      "Name: Mouse, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------\n",
      "Number of rows in the second column: 29\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 1, there are only 5 columns so let's look at all of them\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds3_first_column = ds_3.iloc[:, 0]  \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds3_row_count_1 = ds3_first_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---')\n",
    "print(ds3_first_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_1}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "84ce5c40-8708-4352-ab99-4eb851dfc896",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---\n",
      "0             W\n",
      "1             W\n",
      "2             W\n",
      "3             W\n",
      "4             W\n",
      "5             W\n",
      "6             R\n",
      "7             R\n",
      "8             R\n",
      "9             W\n",
      "10            W\n",
      "11            R\n",
      "12            R\n",
      "13            R\n",
      "14    Treatment\n",
      "15            W\n",
      "16            W\n",
      "17            W\n",
      "18            W\n",
      "19            W\n",
      "20            W\n",
      "21            R\n",
      "22            R\n",
      "23            R\n",
      "24            W\n",
      "25            W\n",
      "26            R\n",
      "27            R\n",
      "28            R\n",
      "Name: Treatment, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------\n",
      "Number of rows in the second column: 29\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 2\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds3_second_column = ds_3.iloc[:, 1]  \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds3_row_count_2 = ds3_second_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN TWO*---')\n",
    "print(ds3_second_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN TWO*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN TWO*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_2}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN TWO*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "68a09aa9-ca7f-4969-b75d-62eef7e37e39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*---\n",
      "0       F\n",
      "1       F\n",
      "2       M\n",
      "3       M\n",
      "4       M\n",
      "5       M\n",
      "6       F\n",
      "7       M\n",
      "8       F\n",
      "9       F\n",
      "10      F\n",
      "11      F\n",
      "12      M\n",
      "13      M\n",
      "14    Sex\n",
      "15      F\n",
      "16      F\n",
      "17      M\n",
      "18      M\n",
      "19      M\n",
      "20      M\n",
      "21      F\n",
      "22      M\n",
      "23      F\n",
      "24      F\n",
      "25      F\n",
      "26      F\n",
      "27      M\n",
      "28      M\n",
      "Name: Sex, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN THREE*------\n",
      "Number of rows in the second column: 29\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN THREE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 3\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds3_third_column = ds_3.iloc[:, 2] \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds3_row_count_3 = ds3_third_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN THREE*---')\n",
    "print(ds3_third_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN THREE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN THREE*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_3}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN THREE*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a4c27ccf-30d0-4a1b-b1cc-0f659dbb20e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN FOUR*---\n",
      "0             34.75\n",
      "1             25.57\n",
      "2             33.91\n",
      "3         37.272731\n",
      "4             28.93\n",
      "5             32.27\n",
      "6             24.22\n",
      "7             23.97\n",
      "8             35.88\n",
      "9             27.38\n",
      "10            31.96\n",
      "11            35.23\n",
      "12            27.98\n",
      "13             28.9\n",
      "14    dCt CRP Liver\n",
      "15                0\n",
      "16             3.56\n",
      "17         3.699234\n",
      "18         1.147269\n",
      "19             0.73\n",
      "20             4.51\n",
      "21         3.635364\n",
      "22             2.62\n",
      "23                0\n",
      "24             2.66\n",
      "25         2.727304\n",
      "26             2.85\n",
      "27         3.964725\n",
      "28         2.456233\n",
      "Name: GADPH Ct Liver, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN FOUR*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN FOUR*------\n",
      "Number of rows in the second column: 29\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN FOUR*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 4\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds3_fourth_column = ds_3.iloc[:, 3]  \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds3_row_count_4 = ds3_fourth_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN FOUR*---')\n",
    "print(ds3_fourth_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN FOUR*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN FOUR*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_4}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN FOUR*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a0187a7e-f0d0-4a5b-b831-9182d5992d39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN FIVE*---\n",
      "0           NaN\n",
      "1     29.130000\n",
      "2     37.609234\n",
      "3     38.420000\n",
      "4     29.660000\n",
      "5     36.780000\n",
      "6     27.855364\n",
      "7     26.590000\n",
      "8           NaN\n",
      "9     30.040000\n",
      "10    34.687304\n",
      "11    38.080000\n",
      "12    31.944725\n",
      "13    31.356233\n",
      "14          NaN\n",
      "15          NaN\n",
      "16          NaN\n",
      "17          NaN\n",
      "18          NaN\n",
      "19          NaN\n",
      "20          NaN\n",
      "21          NaN\n",
      "22          NaN\n",
      "23          NaN\n",
      "24          NaN\n",
      "25          NaN\n",
      "26          NaN\n",
      "27          NaN\n",
      "28          NaN\n",
      "Name: CRP Liver Ct, dtype: float64\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN FIVE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN FIVE*------\n",
      "Number of rows in the second column: 12\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN FIVE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 4\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds3_fifth_column = ds_3.iloc[:, 4]  \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds3_row_count_5 = ds3_fifth_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN FIVE*---')\n",
    "print(ds3_fifth_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN FIVE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN FIVE*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_5}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN FIVE*----')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c0a1560-ddd6-449d-96df-23fcd820d324",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.2 Dataset Three Cleaning</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "542d40b8-b0db-44ef-a0c6-a2797319df3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " Mouse              0\n",
      "Treatment          0\n",
      "Sex                0\n",
      "GADPH Ct Liver     0\n",
      "CRP Liver Ct      17\n",
      "N/A               29\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check for missing values\n",
    "missing_values_3 = ds_3.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "cca598ed-7125-40f7-ad95-8ea32bc16f57",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A\n",
      "0     83         W   F          34.75           NaN  NaN\n",
      "1     85         W   F          25.57     29.130000  NaN\n",
      "2     86         W   M          33.91     37.609234  NaN\n",
      "3     87         W   M      37.272731     38.420000  NaN\n",
      "4     89         W   M          28.93     29.660000  NaN\n",
      "5     90         W   M          32.27     36.780000  NaN\n",
      "6     94         R   F          24.22     27.855364  NaN\n",
      "7     99         R   M          23.97     26.590000  NaN\n",
      "8    249         R   F          35.88           NaN  NaN\n",
      "9     81         W   F          27.38     30.040000  NaN\n",
      "10    82         W   F          31.96     34.687304  NaN\n",
      "11    92         R   F          35.23     38.080000  NaN\n",
      "12   296         R   M          27.98     31.944725  NaN\n"
     ]
    }
   ],
   "source": [
    "# Drop rows 14 and beyond for the first three columns\n",
    "ds_3.iloc[14:, :3] = pd.NA  # Temporarily set rows 14+ in the first three columns to NA\n",
    "\n",
    "# Drop rows with NA values in the first three columns\n",
    "ds_3 = ds_3.dropna(subset=ds_3.columns[:3])\n",
    "\n",
    "# Display the updated DataFrame\n",
    "print(ds_3.head(20))  # Adjust the number of rows displayed as needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "496a15ca-98d8-4784-8973-bc11a9211cb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Series([], Name: GADPH Ct Liver, dtype: object)\n"
     ]
    }
   ],
   "source": [
    "print(ds_3.iloc[14:, 3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "e68ea59f-b28c-47d8-aec9-410ed28f8ef5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A\n",
      "0    83         W   F          34.75           NaN  NaN\n",
      "1    85         W   F          25.57     29.130000  NaN\n",
      "2    86         W   M          33.91     37.609234  NaN\n",
      "3    87         W   M      37.272731     38.420000  NaN\n",
      "4    89         W   M          28.93     29.660000  NaN\n"
     ]
    }
   ],
   "source": [
    "ds_3_original = pd.read_excel(file_path_3)\n",
    "\n",
    "ds_3 = ds_3_original\n",
    "\n",
    "print(ds_3.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "943db97f-9636-4aa0-bada-e8ca659138a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A\n",
      "0    83         W   F          34.75           NaN  NaN\n",
      "1    85         W   F          25.57     29.130000  NaN\n",
      "2    86         W   M          33.91     37.609234  NaN\n",
      "3    87         W   M      37.272731     38.420000  NaN\n",
      "4    89         W   M          28.93     29.660000  NaN\n"
     ]
    }
   ],
   "source": [
    "ds_3_original = pd.read_excel(file_path_3)\n",
    "\n",
    "ds_3 = ds_3_original\n",
    "\n",
    "print(ds_3.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4ea5c225-36bd-4d99-ba8c-57e0aeb6e349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A New Column\n",
      "24    81         W   F           2.66           NaN  NaN       2.66\n",
      "25    82         W   F       2.727304           NaN  NaN   2.727304\n",
      "26    92         R   F           2.85           NaN  NaN       2.85\n",
      "27   296         R   M       3.964725           NaN  NaN   3.964725\n",
      "28   298         R   M       2.456233           NaN  NaN   2.456233\n"
     ]
    }
   ],
   "source": [
    "# Extract values from the fourth column (index 3) starting from index 14\n",
    "new_column_values = ds_3.iloc[14:, 3].reset_index(drop=True)\n",
    "\n",
    "# Create a new column DataFrame with the extracted values\n",
    "# We need to match the length of the new column with the original DataFrame\n",
    "new_column_df = pd.DataFrame({\n",
    "    'New Column': pd.concat([pd.Series([pd.NA] * 14), new_column_values], ignore_index=True)\n",
    "})\n",
    "\n",
    "# Ensure the new column has the same number of rows as the original DataFrame\n",
    "df_new = pd.concat([ds_3, new_column_df], axis=1)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "print(df_new.tail())  #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "987133bf-4f3e-4194-afee-a8b4f19b561a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  N/A New Column\n",
      "0    83         W   F          34.75           NaN  NaN       <NA>\n",
      "1    85         W   F          25.57     29.130000  NaN       <NA>\n",
      "2    86         W   M          33.91     37.609234  NaN       <NA>\n",
      "3    87         W   M      37.272731     38.420000  NaN       <NA>\n",
      "4    89         W   M          28.93     29.660000  NaN       <NA>\n"
     ]
    }
   ],
   "source": [
    "print(df_new.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "c5934c4b-f98e-4623-957c-5393169a298b",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[131], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(df_new\u001b[38;5;241m.\u001b[39miloc[:, \u001b[38;5;241m6\u001b[39m])\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1184\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1182\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1183\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1184\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_tuple(key)\n\u001b[1;32m   1185\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1186\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1187\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1690\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1689\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_getitem_tuple\u001b[39m(\u001b[38;5;28mself\u001b[39m, tup: \u001b[38;5;28mtuple\u001b[39m):\n\u001b[0;32m-> 1690\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_tuple_indexer(tup)\n\u001b[1;32m   1691\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1692\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_lowerdim(tup)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:966\u001b[0m, in \u001b[0;36m_LocationIndexer._validate_tuple_indexer\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    964\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, k \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(key):\n\u001b[1;32m    965\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 966\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(k, i)\n\u001b[1;32m    967\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    968\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    969\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLocation based indexing can only have \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    970\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_valid_types\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] types\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    971\u001b[0m         ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1592\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1590\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m   1591\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_integer(key):\n\u001b[0;32m-> 1592\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_integer(key, axis)\n\u001b[1;32m   1593\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m   1594\u001b[0m     \u001b[38;5;66;03m# a tuple should already have been caught by this point\u001b[39;00m\n\u001b[1;32m   1595\u001b[0m     \u001b[38;5;66;03m# so don't treat a tuple as a valid indexer\u001b[39;00m\n\u001b[1;32m   1596\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m IndexingError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mToo many indexers\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1685\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_integer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1683\u001b[0m len_axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis(axis))\n\u001b[1;32m   1684\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m len_axis \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m-\u001b[39mlen_axis:\n\u001b[0;32m-> 1685\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msingle positional indexer is out-of-bounds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "print(df_new.iloc[:, 6])\n",
    "\n",
    "#IndexError: single positional indexer is out-of-bounds because data is allready dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "554ab51c-38bd-4e3f-9bd9-185959d79903",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[124], line 59\u001b[0m\n\u001b[1;32m     55\u001b[0m df_new\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m27\u001b[39m, \u001b[38;5;241m5\u001b[39m] \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mNA\n\u001b[1;32m     57\u001b[0m df_new \u001b[38;5;241m=\u001b[39m df_new\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{df_new\u001b[38;5;241m.\u001b[39mcolumns[\u001b[38;5;241m5\u001b[39m]: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdCt CRP Liver\u001b[39m\u001b[38;5;124m'\u001b[39m})\n\u001b[0;32m---> 59\u001b[0m \u001b[38;5;28mprint\u001b[39m(df_new\u001b[38;5;241m.\u001b[39miloc[:, \u001b[38;5;241m6\u001b[39m])\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1184\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1182\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_scalar_access(key):\n\u001b[1;32m   1183\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_value(\u001b[38;5;241m*\u001b[39mkey, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n\u001b[0;32m-> 1184\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_tuple(key)\n\u001b[1;32m   1185\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1186\u001b[0m     \u001b[38;5;66;03m# we by definition only have the 0th axis\u001b[39;00m\n\u001b[1;32m   1187\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1690\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1689\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_getitem_tuple\u001b[39m(\u001b[38;5;28mself\u001b[39m, tup: \u001b[38;5;28mtuple\u001b[39m):\n\u001b[0;32m-> 1690\u001b[0m     tup \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_tuple_indexer(tup)\n\u001b[1;32m   1691\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m suppress(IndexingError):\n\u001b[1;32m   1692\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_lowerdim(tup)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:966\u001b[0m, in \u001b[0;36m_LocationIndexer._validate_tuple_indexer\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    964\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, k \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(key):\n\u001b[1;32m    965\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 966\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(k, i)\n\u001b[1;32m    967\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    968\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    969\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLocation based indexing can only have \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    970\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_valid_types\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] types\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    971\u001b[0m         ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1592\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_key\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1590\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m   1591\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_integer(key):\n\u001b[0;32m-> 1592\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_integer(key, axis)\n\u001b[1;32m   1593\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m   1594\u001b[0m     \u001b[38;5;66;03m# a tuple should already have been caught by this point\u001b[39;00m\n\u001b[1;32m   1595\u001b[0m     \u001b[38;5;66;03m# so don't treat a tuple as a valid indexer\u001b[39;00m\n\u001b[1;32m   1596\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m IndexingError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mToo many indexers\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexing.py:1685\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_integer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1683\u001b[0m len_axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis(axis))\n\u001b[1;32m   1684\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m len_axis \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m-\u001b[39mlen_axis:\n\u001b[0;32m-> 1685\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msingle positional indexer is out-of-bounds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "df_new.iloc[0, 5] = df_new.iloc[15, 3]\n",
    "\n",
    "df_new.iloc[1, 5] = df_new.iloc[16, 3]\n",
    "\n",
    "df_new.iloc[2, 5] = df_new.iloc[17, 3]\n",
    "\n",
    "df_new.iloc[3, 5] = df_new.iloc[18, 3]\n",
    "\n",
    "df_new.iloc[4, 5] = df_new.iloc[19, 3]\n",
    "\n",
    "df_new.iloc[5, 5] = df_new.iloc[20, 3]\n",
    "\n",
    "df_new.iloc[6, 5] = df_new.iloc[21, 3]\n",
    "\n",
    "df_new.iloc[7, 5] = df_new.iloc[22, 3]\n",
    "\n",
    "df_new.iloc[8, 5] = df_new.iloc[23, 3]\n",
    "\n",
    "df_new.iloc[9, 5] = df_new.iloc[24, 3]\n",
    "\n",
    "df_new.iloc[10, 5] = df_new.iloc[25, 3]\n",
    "\n",
    "df_new.iloc[11, 5] = df_new.iloc[26, 3]\n",
    "\n",
    "df_new.iloc[12, 5] = df_new.iloc[27, 3]\n",
    "\n",
    "df_new.iloc[13, 5] = df_new.iloc[28, 3]\n",
    "\n",
    "df_new.iloc[14, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[15, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[16, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[17, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[18, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[19, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[20, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[21, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[22, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[23, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[24, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[25, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[26, 5] = pd.NA\n",
    "\n",
    "df_new.iloc[27, 5] = pd.NA\n",
    "\n",
    "df_new = df_new.rename(columns={df_new.columns[5]: 'dCt CRP Liver'})\n",
    "\n",
    "print(df_new.iloc[:, 6])\n",
    "\n",
    "#IndexError: single positional indexer is out-of-bounds because data is allready dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "412dc823-2ad6-4d6c-b989-1b46b85e11ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Mouse  Treatment  Sex GADPH Ct Liver  CRP Liver Ct  dCt CRP Liver  \\\n",
      "0      83          W    F          34.75           NaN       0.000000   \n",
      "1      85          W    F          25.57     29.130000       3.560000   \n",
      "2      86          W    M          33.91     37.609234       3.699234   \n",
      "3      87          W    M      37.272731     38.420000       1.147269   \n",
      "4      89          W    M          28.93     29.660000       0.730000   \n",
      "5      90          W    M          32.27     36.780000       4.510000   \n",
      "6      94          R    F          24.22     27.855364       3.635364   \n",
      "7      99          R    M          23.97     26.590000       2.620000   \n",
      "8     249          R    F          35.88           NaN       0.000000   \n",
      "9      81          W    F          27.38     30.040000       2.660000   \n",
      "10     82          W    F          31.96     34.687304       2.727304   \n",
      "11     92          R    F          35.23     38.080000       2.850000   \n",
      "12    296          R    M          27.98     31.944725       3.964725   \n",
      "13    298          R    M           28.9     31.356233       2.456233   \n",
      "14  Mouse  Treatment  Sex  dCt CRP Liver           NaN            NaN   \n",
      "15     83          W    F              0           NaN            NaN   \n",
      "16     85          W    F           3.56           NaN            NaN   \n",
      "17     86          W    M       3.699234           NaN            NaN   \n",
      "18     87          W    M       1.147269           NaN            NaN   \n",
      "19     89          W    M           0.73           NaN            NaN   \n",
      "20     90          W    M           4.51           NaN            NaN   \n",
      "21     94          R    F       3.635364           NaN            NaN   \n",
      "22     99          R    M           2.62           NaN            NaN   \n",
      "23    249          R    F              0           NaN            NaN   \n",
      "24     81          W    F           2.66           NaN            NaN   \n",
      "25     82          W    F       2.727304           NaN            NaN   \n",
      "26     92          R    F           2.85           NaN            NaN   \n",
      "27    296          R    M       3.964725           NaN            NaN   \n",
      "28    298          R    M       2.456233           NaN            NaN   \n",
      "\n",
      "   New Column  \n",
      "0         NaN  \n",
      "1         NaN  \n",
      "2         NaN  \n",
      "3         NaN  \n",
      "4         NaN  \n",
      "5         NaN  \n",
      "6         NaN  \n",
      "7         NaN  \n",
      "8         NaN  \n",
      "9         NaN  \n",
      "10        NaN  \n",
      "11        NaN  \n",
      "12        NaN  \n",
      "13        NaN  \n",
      "14       <NA>  \n",
      "15       <NA>  \n",
      "16       <NA>  \n",
      "17       <NA>  \n",
      "18       <NA>  \n",
      "19       <NA>  \n",
      "20       <NA>  \n",
      "21       <NA>  \n",
      "22       <NA>  \n",
      "23       <NA>  \n",
      "24       <NA>  \n",
      "25       <NA>  \n",
      "26       <NA>  \n",
      "27       <NA>  \n",
      "28       <NA>  \n"
     ]
    }
   ],
   "source": [
    "print(df_new[:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "a6e60493-7f44-4350-9196-34b8031bf968",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['New Column'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[126], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df_new \u001b[38;5;241m=\u001b[39m df_new\u001b[38;5;241m.\u001b[39mdrop(columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNew Column\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(df_new[:])\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/frame.py:5581\u001b[0m, in \u001b[0;36mDataFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   5433\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdrop\u001b[39m(\n\u001b[1;32m   5434\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   5435\u001b[0m     labels: IndexLabel \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5442\u001b[0m     errors: IgnoreRaise \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraise\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   5443\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   5444\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5445\u001b[0m \u001b[38;5;124;03m    Drop specified labels from rows or columns.\u001b[39;00m\n\u001b[1;32m   5446\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5579\u001b[0m \u001b[38;5;124;03m            weight  1.0     0.8\u001b[39;00m\n\u001b[1;32m   5580\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 5581\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mdrop(\n\u001b[1;32m   5582\u001b[0m         labels\u001b[38;5;241m=\u001b[39mlabels,\n\u001b[1;32m   5583\u001b[0m         axis\u001b[38;5;241m=\u001b[39maxis,\n\u001b[1;32m   5584\u001b[0m         index\u001b[38;5;241m=\u001b[39mindex,\n\u001b[1;32m   5585\u001b[0m         columns\u001b[38;5;241m=\u001b[39mcolumns,\n\u001b[1;32m   5586\u001b[0m         level\u001b[38;5;241m=\u001b[39mlevel,\n\u001b[1;32m   5587\u001b[0m         inplace\u001b[38;5;241m=\u001b[39minplace,\n\u001b[1;32m   5588\u001b[0m         errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m   5589\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/generic.py:4788\u001b[0m, in \u001b[0;36mNDFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4786\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m axis, labels \u001b[38;5;129;01min\u001b[39;00m axes\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m   4787\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m labels \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 4788\u001b[0m         obj \u001b[38;5;241m=\u001b[39m obj\u001b[38;5;241m.\u001b[39m_drop_axis(labels, axis, level\u001b[38;5;241m=\u001b[39mlevel, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[1;32m   4790\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[1;32m   4791\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_inplace(obj)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/generic.py:4830\u001b[0m, in \u001b[0;36mNDFrame._drop_axis\u001b[0;34m(self, labels, axis, level, errors, only_slice)\u001b[0m\n\u001b[1;32m   4828\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mdrop(labels, level\u001b[38;5;241m=\u001b[39mlevel, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[1;32m   4829\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4830\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mdrop(labels, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[1;32m   4831\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mget_indexer(new_axis)\n\u001b[1;32m   4833\u001b[0m \u001b[38;5;66;03m# Case for non-unique axis\u001b[39;00m\n\u001b[1;32m   4834\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/indexes/base.py:7070\u001b[0m, in \u001b[0;36mIndex.drop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   7068\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mask\u001b[38;5;241m.\u001b[39many():\n\u001b[1;32m   7069\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m errors \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 7070\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlabels[mask]\u001b[38;5;241m.\u001b[39mtolist()\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not found in axis\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   7071\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m indexer[\u001b[38;5;241m~\u001b[39mmask]\n\u001b[1;32m   7072\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdelete(indexer)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['New Column'] not found in axis\""
     ]
    }
   ],
   "source": [
    "df_new = df_new.drop(columns=['New Column'])\n",
    "print(df_new[:])\n",
    "\n",
    "#KeyError: \"['New Column'] not found in axis\" because column is already dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "f16e0fe9-9cbe-4316-8740-387b52793e74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  dCt CRP Liver\n",
      "0     83         W   F          34.75           NaN       0.000000\n",
      "1     85         W   F          25.57     29.130000       3.560000\n",
      "2     86         W   M          33.91     37.609234       3.699234\n",
      "3     87         W   M      37.272731     38.420000       1.147269\n",
      "4     89         W   M          28.93     29.660000       0.730000\n",
      "5     90         W   M          32.27     36.780000       4.510000\n",
      "6     94         R   F          24.22     27.855364       3.635364\n",
      "7     99         R   M          23.97     26.590000       2.620000\n",
      "8    249         R   F          35.88           NaN       0.000000\n",
      "9     81         W   F          27.38     30.040000       2.660000\n",
      "10    82         W   F          31.96     34.687304       2.727304\n",
      "11    92         R   F          35.23     38.080000       2.850000\n",
      "12   296         R   M          27.98     31.944725       3.964725\n",
      "13   298         R   M           28.9     31.356233       2.456233\n"
     ]
    }
   ],
   "source": [
    "ds_3 = df_new.drop(df_new.index[14:29])\n",
    "print(ds_3[:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "1504195b-9737-49c7-9dd2-24ca20af0f25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Mouse Treatment Sex GADPH Ct Liver  CRP Liver Ct  dCt CRP Liver\n",
      "0     83         W   F          34.75      0.000000       0.000000\n",
      "1     85         W   F          25.57     29.130000       3.560000\n",
      "2     86         W   M          33.91     37.609234       3.699234\n",
      "3     87         W   M      37.272731     38.420000       1.147269\n",
      "4     89         W   M          28.93     29.660000       0.730000\n",
      "5     90         W   M          32.27     36.780000       4.510000\n",
      "6     94         R   F          24.22     27.855364       3.635364\n",
      "7     99         R   M          23.97     26.590000       2.620000\n",
      "8    249         R   F          35.88      0.000000       0.000000\n",
      "9     81         W   F          27.38     30.040000       2.660000\n",
      "10    82         W   F          31.96     34.687304       2.727304\n",
      "11    92         R   F          35.23     38.080000       2.850000\n",
      "12   296         R   M          27.98     31.944725       3.964725\n",
      "13   298         R   M           28.9     31.356233       2.456233\n"
     ]
    }
   ],
   "source": [
    "ds_3.iloc[0, 4] = 0.00\n",
    "\n",
    "ds_3.iloc[8, 4] = 0.00\n",
    "\n",
    "print(ds_3[:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "736d95bc-fc36-4cf6-985f-90a2e6bf1e4c",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.6.1 Dataset 3 Key Concepts</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "**Rhodiola Rosea:** Is a herb that is found at high altitudes, in cold weather. It is primarily known as an adaptogen, which are substances that help someone adapt to stress by lowering stress and cortisol. There are other therapeutic benefits to this herb such as improving biomarkers in gut microbiota within the study that is this dataset<a href=\"#mla_151\"> [151]</a>.\n",
    "\n",
    "<br><blockquote style=\"font-style: italic; margin: 20px; border-left: 4px solid #ccc; padding-left: 15px;\">Rhodiola Rosea lowers inflammation within the GI Tract, thus improving gene expression in this study<a href=\"#mla_152\"> [152]</a>.</blockquote><br>\n",
    "\n",
    "\n",
    "**CRP:** C-Reactive Protien is a protein released by the liver in response to systemic inflammation. Lower levels indicates less inflammation<a href=\"#mla_153\"> [153]</a>.\n",
    "\n",
    "**GAPDH (Glyceraldehyde 3-phosphate dehydrogenase):** An enzyme that breaks down glucose for energy, perfect to use as a reference gene<a href=\"#mla_154\"> [154]</a>.\n",
    "\n",
    "**Ct (Cycle threshold):** Used in PCR for DNA analysis. A lower value means a higher amount of the target gene, and a higher value means a lower amount of the target gene<a href=\"#mla_155\"> [155]</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb4c60ec-b313-4cb3-98dc-fa836943c8f1",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.6.2 Dataset 3 Abstract</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "\"Given the limitations and side effects associated with pharmaceuticals currently being used for therapy of diabetes, there is a significant need for alternative treatments. In this study, we investigated the effects of a root extract from Rhodiola rosea in a Leptin receptor knockout (db/db) mouse model of type 2 diabetes\". Leptin signals to the brain to reduce hunger so in a leptin receptor knockout model some mouse groups were always hungry<a href=\"#mla_148\"> [148]</a>.\n",
    "\n",
    "\"Rhodiola rosea improved fasting blood glucose levels, altered the response to exogenous insulin, and decreased circulating lipopolysaccharide and hepatic C-reactive protein transcript levels. We hypothesize that these changes may in part reflect the modulation of the microbiota, resulting in improved gut barrier integrity and decreasing the translocation of inflammatory biomolecules into the bloodstream\"<a href=\"#mla_149\"> [149]</a>.\n",
    "\n",
    "There were two treatment groups, water-treated mice, and Rhodiola-treated mice. The CRP (gene of interest) and GAPDH (reference gene) were examined from liver extracts of the mice and were processed using DNA tests and a \"PCR amplification program\"<a href=\"#mla_150\"> [150]</a>.\n",
    "\n",
    "<h4>Dataset 3 (columns):</h4>  \n",
    "\n",
    "- 1.) Mouse: Specimen.\n",
    "- 2.) Treatment: Rhodiola vs. water.\n",
    "- 3.) Sex: Gender.\n",
    "- 4.) GADPH Ct Liver: Reference Gene.\n",
    "- 5.) CRP Liver CT: Gene of Interest.\n",
    "\n",
    "<h4>Conclusion</h4>\n",
    "\n",
    "\n",
    "All five columns describe the information about the mouse specimen, and two columns on a reference gene, and gene of interest from the sample data. This dataset aims to view the concentrations of each reference gene and each gene of interest in order to see how each gene is expressed based on whether the mouse was in the Rhodiola group or the control water group. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7027b97e-7794-4fbf-b301-80080baa1780",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.7 Dataset Four</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d726f22-a6f5-431b-9881-d27e9217cedf",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h3>1.3.8 Dataset Four Overview</h3>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "dc29037b-884d-4f6f-9141-8fa342283470",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "--------------------*INFO*----------------------\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 35557 entries, 0 to 35556\n",
      "Data columns (total 27 columns):\n",
      " #   Column      Non-Null Count  Dtype  \n",
      "---  ------      --------------  -----  \n",
      " 0   ID_REF      35557 non-null  object \n",
      " 1   GSM3101204  35556 non-null  float64\n",
      " 2   GSM3101205  35556 non-null  float64\n",
      " 3   GSM3101206  35556 non-null  float64\n",
      " 4   GSM3101207  35556 non-null  float64\n",
      " 5   GSM3101208  35556 non-null  float64\n",
      " 6   GSM3101209  35556 non-null  float64\n",
      " 7   GSM3101210  35556 non-null  float64\n",
      " 8   GSM3101211  35556 non-null  float64\n",
      " 9   GSM3101212  35556 non-null  float64\n",
      " 10  GSM3101213  35556 non-null  float64\n",
      " 11  GSM3101214  35556 non-null  float64\n",
      " 12  GSM3101215  35556 non-null  float64\n",
      " 13  GSM3101216  35556 non-null  float64\n",
      " 14  GSM3101217  35556 non-null  float64\n",
      " 15  GSM3101218  35556 non-null  float64\n",
      " 16  GSM3101219  35556 non-null  float64\n",
      " 17  GSM3101220  35556 non-null  float64\n",
      " 18  GSM3101221  35556 non-null  float64\n",
      " 19  GSM3101222  35556 non-null  float64\n",
      " 20  GSM3101223  35556 non-null  float64\n",
      " 21  GSM3101224  35556 non-null  float64\n",
      " 22  GSM3101225  35556 non-null  float64\n",
      " 23  GSM3101226  35556 non-null  float64\n",
      " 24  GSM3101227  35556 non-null  float64\n",
      " 25  GSM3101228  35556 non-null  float64\n",
      " 26  GSM3101229  35556 non-null  float64\n",
      "dtypes: float64(26), object(1)\n",
      "memory usage: 7.3+ MB\n",
      "None\n",
      "------------------*END INFO*--------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*COLUMNS*--------------------\n",
      "Index(['ID_REF', 'GSM3101204', 'GSM3101205', 'GSM3101206', 'GSM3101207',\n",
      "       'GSM3101208', 'GSM3101209', 'GSM3101210', 'GSM3101211', 'GSM3101212',\n",
      "       'GSM3101213', 'GSM3101214', 'GSM3101215', 'GSM3101216', 'GSM3101217',\n",
      "       'GSM3101218', 'GSM3101219', 'GSM3101220', 'GSM3101221', 'GSM3101222',\n",
      "       'GSM3101223', 'GSM3101224', 'GSM3101225', 'GSM3101226', 'GSM3101227',\n",
      "       'GSM3101228', 'GSM3101229'],\n",
      "      dtype='object')\n",
      "-----------------*END COLUMNS*------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------------------*DATA TYPES*------------------\n",
      "ID_REF         object\n",
      "GSM3101204    float64\n",
      "GSM3101205    float64\n",
      "GSM3101206    float64\n",
      "GSM3101207    float64\n",
      "GSM3101208    float64\n",
      "GSM3101209    float64\n",
      "GSM3101210    float64\n",
      "GSM3101211    float64\n",
      "GSM3101212    float64\n",
      "GSM3101213    float64\n",
      "GSM3101214    float64\n",
      "GSM3101215    float64\n",
      "GSM3101216    float64\n",
      "GSM3101217    float64\n",
      "GSM3101218    float64\n",
      "GSM3101219    float64\n",
      "GSM3101220    float64\n",
      "GSM3101221    float64\n",
      "GSM3101222    float64\n",
      "GSM3101223    float64\n",
      "GSM3101224    float64\n",
      "GSM3101225    float64\n",
      "GSM3101226    float64\n",
      "GSM3101227    float64\n",
      "GSM3101228    float64\n",
      "GSM3101229    float64\n",
      "dtype: object\n",
      "----------------*END DATA TYPES*----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "---------------------*HEAD*---------------------\n",
      "     ID_REF  GSM3101204  GSM3101205  GSM3101206  GSM3101207  GSM3101208  \\\n",
      "0  10338001   11.621818   11.723674   11.797580   11.845637   11.706340   \n",
      "1  10338002    4.882116    4.819835    4.945165    4.816617    5.066188   \n",
      "2  10338003    9.815821    9.913645   10.070279   10.068832    9.953928   \n",
      "3  10338004    9.123786    9.135206    9.285084    9.463849    9.290599   \n",
      "4  10338005    2.294406    2.362550    2.367593    2.393940    2.387859   \n",
      "\n",
      "   GSM3101209  GSM3101210  GSM3101211  GSM3101212  ...  GSM3101220  \\\n",
      "0   11.619102   11.640126   11.753207   11.671719  ...   11.662705   \n",
      "1    5.032585    4.809279    4.929813    4.989135  ...    4.954034   \n",
      "2    9.841130    9.849367    9.990176    9.907970  ...    9.884320   \n",
      "3    9.101129    9.287354    9.266923    9.084664  ...    9.149640   \n",
      "4    2.398327    2.437759    2.382819    2.403590  ...    2.363132   \n",
      "\n",
      "   GSM3101221  GSM3101222  GSM3101223  GSM3101224  GSM3101225  GSM3101226  \\\n",
      "0   11.777588   11.689259   11.636659   11.690642   11.611120   11.738238   \n",
      "1    4.783552    4.968062    5.006434    4.920746    4.877134    4.843854   \n",
      "2    9.921198    9.963101    9.855835    9.898591    9.785410    9.955079   \n",
      "3    9.297222    9.093457    9.064904    8.989178    9.049290    9.150534   \n",
      "4    2.345471    2.368008    2.357114    2.355657    2.301924    2.357530   \n",
      "\n",
      "   GSM3101227  GSM3101228  GSM3101229  \n",
      "0   11.653844   11.635455   11.663165  \n",
      "1    4.951407    4.874028    4.746962  \n",
      "2    9.833608    9.770072    9.802227  \n",
      "3    9.144118    8.987364    9.074741  \n",
      "4    2.324505    2.324368    2.361027  \n",
      "\n",
      "[5 rows x 27 columns]\n",
      "-------------------*END HEAD*-------------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-------------------*DESCRIBE*-------------------\n",
      "         GSM3101204    GSM3101205    GSM3101206    GSM3101207    GSM3101208  \\\n",
      "count  35556.000000  35556.000000  35556.000000  35556.000000  35556.000000   \n",
      "mean       6.590190      6.600400      6.577663      6.601053      6.598796   \n",
      "std        2.410057      2.411271      2.404316      2.412906      2.394592   \n",
      "min        1.901215      1.913130      1.797553      1.916115      1.874138   \n",
      "25%        4.674652      4.646850      4.637177      4.665972      4.709954   \n",
      "50%        6.300616      6.331360      6.319351      6.272808      6.271094   \n",
      "75%        8.366900      8.379242      8.337251      8.405842      8.368564   \n",
      "max       14.093398     14.161691     14.200280     14.144030     14.090882   \n",
      "\n",
      "         GSM3101209    GSM3101210    GSM3101211    GSM3101212    GSM3101213  \\\n",
      "count  35556.000000  35556.000000  35556.000000  35556.000000  35556.000000   \n",
      "mean       6.599403      6.601156      6.576549      6.591168      6.596996   \n",
      "std        2.403573      2.401831      2.398243      2.402301      2.409805   \n",
      "min        1.736534      1.739100      1.838965      1.663046      1.795649   \n",
      "25%        4.696579      4.680109      4.653438      4.682758      4.673260   \n",
      "50%        6.269725      6.284596      6.300395      6.293038      6.307941   \n",
      "75%        8.390788      8.377255      8.331029      8.365709      8.370828   \n",
      "max       14.122711     14.138157     14.136341     14.111863     14.008937   \n",
      "\n",
      "       ...    GSM3101220    GSM3101221    GSM3101222    GSM3101223  \\\n",
      "count  ...  35556.000000  35556.000000  35556.000000  35556.000000   \n",
      "mean   ...      6.585819      6.589362      6.586782      6.582773   \n",
      "std    ...      2.409562      2.412676      2.399346      2.397149   \n",
      "min    ...      1.814080      1.873293      1.810988      1.932036   \n",
      "25%    ...      4.674540      4.647733      4.683279      4.677987   \n",
      "50%    ...      6.295058      6.311179      6.288637      6.300732   \n",
      "75%    ...      8.365994      8.364780      8.352357      8.348847   \n",
      "max    ...     14.124468     14.150732     14.094836     14.129696   \n",
      "\n",
      "         GSM3101224    GSM3101225    GSM3101226    GSM3101227    GSM3101228  \\\n",
      "count  35556.000000  35556.000000  35556.000000  35556.000000  35556.000000   \n",
      "mean       6.571828      6.579734      6.580785      6.583967      6.589516   \n",
      "std        2.405675      2.412536      2.403029      2.410884      2.415121   \n",
      "min        1.735129      1.888741      1.739100      1.753590      1.814460   \n",
      "25%        4.637038      4.658735      4.664986      4.666023      4.649021   \n",
      "50%        6.296382      6.296361      6.284290      6.290051      6.323170   \n",
      "75%        8.335034      8.341868      8.341906      8.357242      8.371671   \n",
      "max       14.148078     14.079068     14.100092     14.166087     14.159905   \n",
      "\n",
      "         GSM3101229  \n",
      "count  35556.000000  \n",
      "mean       6.590777  \n",
      "std        2.419820  \n",
      "min        1.816958  \n",
      "25%        4.645037  \n",
      "50%        6.297717  \n",
      "75%        8.380687  \n",
      "max       14.165007  \n",
      "\n",
      "[8 rows x 26 columns]\n",
      "-----------------*END DESCRIBE*-----------------\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "----------*END OVERVIEW OF DATASET 3*-----------\n"
     ]
    }
   ],
   "source": [
    "# Overview of Dataset 1\n",
    "\n",
    "# Get basic information about the dataset\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "print('--------------------*INFO*----------------------')\n",
    "print(ds_4.info())\n",
    "print('------------------*END INFO*--------------------')\n",
    "\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# View column names\n",
    "print('-------------------*COLUMNS*--------------------')\n",
    "print(ds_4.columns)\n",
    "print('-----------------*END COLUMNS*------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# List all the columns and their data types\n",
    "print('------------------*DATA TYPES*------------------')\n",
    "print(ds_4.dtypes) \n",
    "print('----------------*END DATA TYPES*----------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Display the first few rows to inspect the data\n",
    "print('---------------------*HEAD*---------------------')\n",
    "print(ds_4.head())\n",
    "print('-------------------*END HEAD*-------------------')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "# Check the summary statistics of numeric columns\n",
    "print('-------------------*DESCRIBE*-------------------')\n",
    "print(ds_4.describe())\n",
    "print('-----------------*END DESCRIBE*-----------------')\n",
    "print('\\n\\n\\n')\n",
    "print('----------*END OVERVIEW OF DATASET 3*-----------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "44bd9b8b-2897-4ee2-a311-993d06991f72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------*BEGIN OVERVIEW OF DATASET 3*----------\n",
      "---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---\n",
      "0                        10338001\n",
      "1                        10338002\n",
      "2                        10338003\n",
      "3                        10338004\n",
      "4                        10338005\n",
      "                   ...           \n",
      "35552                    10608721\n",
      "35553                    10608722\n",
      "35554                    10608723\n",
      "35555                    10608724\n",
      "35556    !series_matrix_table_end\n",
      "Name: ID_REF, Length: 35557, dtype: object\n",
      "\n",
      "\n",
      "-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------\n",
      "Number of rows in the second column: 29\n",
      "----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----\n"
     ]
    }
   ],
   "source": [
    "# Let's look at the number of rows in column 1 it seams of interest.\n",
    "\n",
    "print('---------*BEGIN OVERVIEW OF DATASET 3*----------')\n",
    "\n",
    "# Accessing the second column by its position\n",
    "ds4_first_column = ds_4.iloc[:, 0]  \n",
    "\n",
    "# Counting the number of rows (which is the same as counting non-NaN values)\n",
    "ds4_row_count_1 = ds4_first_column.count()\n",
    "\n",
    "\n",
    "\n",
    "print('---*INSPECTION OF NUMBER OF ROWS :COLUMN ONE*---')\n",
    "print(ds4_first_column)\n",
    "print('\\n')\n",
    "print('-*END INSPECTION OF NUMBER OF ROWS :COLUMN ONE*-')\n",
    "print('\\n\\n\\n\\n\\n')\n",
    "\n",
    "\n",
    "print('------*TOTAL NUMBER OF ROWS OF COLUMN ONE*------')\n",
    "print(f\"Number of rows in the second column: {ds3_row_count_4}\")\n",
    "print('----*END TOTAL NUMBER OF ROWS OF COLUMN ONE*----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "e7efa20a-7431-45dd-b334-d98045e4d8cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " ID_REF        0\n",
      "GSM3101204    1\n",
      "GSM3101205    1\n",
      "GSM3101206    1\n",
      "GSM3101207    1\n",
      "GSM3101208    1\n",
      "GSM3101209    1\n",
      "GSM3101210    1\n",
      "GSM3101211    1\n",
      "GSM3101212    1\n",
      "GSM3101213    1\n",
      "GSM3101214    1\n",
      "GSM3101215    1\n",
      "GSM3101216    1\n",
      "GSM3101217    1\n",
      "GSM3101218    1\n",
      "GSM3101219    1\n",
      "GSM3101220    1\n",
      "GSM3101221    1\n",
      "GSM3101222    1\n",
      "GSM3101223    1\n",
      "GSM3101224    1\n",
      "GSM3101225    1\n",
      "GSM3101226    1\n",
      "GSM3101227    1\n",
      "GSM3101228    1\n",
      "GSM3101229    1\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check for missing values\n",
    "missing_values_4 = ds_4.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7c622277-a29d-4be5-a04e-2b6838d40318",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     ID_REF  GSM3101204  GSM3101205  GSM3101206  GSM3101207  GSM3101208  \\\n",
      "0  10338001   11.621818   11.723674   11.797580   11.845637   11.706340   \n",
      "1  10338002    4.882116    4.819835    4.945165    4.816617    5.066188   \n",
      "2  10338003    9.815821    9.913645   10.070279   10.068832    9.953928   \n",
      "3  10338004    9.123786    9.135206    9.285084    9.463849    9.290599   \n",
      "4  10338005    2.294406    2.362550    2.367593    2.393940    2.387859   \n",
      "\n",
      "   GSM3101209  GSM3101210  GSM3101211  GSM3101212  ...  GSM3101220  \\\n",
      "0   11.619102   11.640126   11.753207   11.671719  ...   11.662705   \n",
      "1    5.032585    4.809279    4.929813    4.989135  ...    4.954034   \n",
      "2    9.841130    9.849367    9.990176    9.907970  ...    9.884320   \n",
      "3    9.101129    9.287354    9.266923    9.084664  ...    9.149640   \n",
      "4    2.398327    2.437759    2.382819    2.403590  ...    2.363132   \n",
      "\n",
      "   GSM3101221  GSM3101222  GSM3101223  GSM3101224  GSM3101225  GSM3101226  \\\n",
      "0   11.777588   11.689259   11.636659   11.690642   11.611120   11.738238   \n",
      "1    4.783552    4.968062    5.006434    4.920746    4.877134    4.843854   \n",
      "2    9.921198    9.963101    9.855835    9.898591    9.785410    9.955079   \n",
      "3    9.297222    9.093457    9.064904    8.989178    9.049290    9.150534   \n",
      "4    2.345471    2.368008    2.357114    2.355657    2.301924    2.357530   \n",
      "\n",
      "   GSM3101227  GSM3101228  GSM3101229  \n",
      "0   11.653844   11.635455   11.663165  \n",
      "1    4.951407    4.874028    4.746962  \n",
      "2    9.833608    9.770072    9.802227  \n",
      "3    9.144118    8.987364    9.074741  \n",
      "4    2.324505    2.324368    2.361027  \n",
      "\n",
      "[5 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "print(ds_4.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "a050fae3-3e9f-4c59-bd6c-123de01cf246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         ID_REF  GSM3101204  GSM3101205  GSM3101206  \\\n",
      "35552                  10608721    7.245398    7.204073    7.169800   \n",
      "35553                  10608722    3.144329    3.171411    3.072336   \n",
      "35554                  10608723    3.402977    3.382930    3.376089   \n",
      "35555                  10608724    4.817116    4.829814    4.971051   \n",
      "35556  !series_matrix_table_end         NaN         NaN         NaN   \n",
      "\n",
      "       GSM3101207  GSM3101208  GSM3101209  GSM3101210  GSM3101211  GSM3101212  \\\n",
      "35552    7.253484    7.661509    7.570066    7.656295    7.337315    7.323211   \n",
      "35553    3.319842    3.280265    3.074154    3.304790    3.208513    3.117621   \n",
      "35554    3.381280    3.401091    3.443200    3.291308    3.356121    3.416632   \n",
      "35555    4.705429    4.764004    4.737668    4.735075    4.813959    4.699708   \n",
      "35556         NaN         NaN         NaN         NaN         NaN         NaN   \n",
      "\n",
      "       ...  GSM3101220  GSM3101221  GSM3101222  GSM3101223  GSM3101224  \\\n",
      "35552  ...    7.371603    7.124561    7.290055    7.296061    7.149179   \n",
      "35553  ...    3.236720    3.134968    3.256094    3.179739    3.188076   \n",
      "35554  ...    3.392597    3.291738    3.392518    3.354275    3.409771   \n",
      "35555  ...    4.805977    4.899230    4.727363    4.816136    4.817920   \n",
      "35556  ...         NaN         NaN         NaN         NaN         NaN   \n",
      "\n",
      "       GSM3101225  GSM3101226  GSM3101227  GSM3101228  GSM3101229  \n",
      "35552    7.215749    7.196333    7.275178    7.244657    6.918887  \n",
      "35553    3.144330    3.207743    3.153569    3.158315    3.186455  \n",
      "35554    3.488536    3.471732    3.400055    3.332553    3.312330  \n",
      "35555    4.805248    4.840851    4.777877    4.794845    4.739567  \n",
      "35556         NaN         NaN         NaN         NaN         NaN  \n",
      "\n",
      "[5 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "print(ds_4.tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "769414d9-980f-4300-9d4a-af3d376c61ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         ID_REF  GSM3101204  GSM3101205  GSM3101206  GSM3101207  GSM3101208  \\\n",
      "35551  10608720    7.992098    7.947808    8.116393    7.858657    7.685737   \n",
      "35552  10608721    7.245398    7.204073    7.169800    7.253484    7.661509   \n",
      "35553  10608722    3.144329    3.171411    3.072336    3.319842    3.280265   \n",
      "35554  10608723    3.402977    3.382930    3.376089    3.381280    3.401091   \n",
      "35555  10608724    4.817116    4.829814    4.971051    4.705429    4.764004   \n",
      "\n",
      "       GSM3101209  GSM3101210  GSM3101211  GSM3101212  ...  GSM3101220  \\\n",
      "35551    7.914463    7.718136    7.673214    7.600938  ...    7.885143   \n",
      "35552    7.570066    7.656295    7.337315    7.323211  ...    7.371603   \n",
      "35553    3.074154    3.304790    3.208513    3.117621  ...    3.236720   \n",
      "35554    3.443200    3.291308    3.356121    3.416632  ...    3.392597   \n",
      "35555    4.737668    4.735075    4.813959    4.699708  ...    4.805977   \n",
      "\n",
      "       GSM3101221  GSM3101222  GSM3101223  GSM3101224  GSM3101225  GSM3101226  \\\n",
      "35551    7.985984    7.909240    7.864568    7.967580    7.929491    7.877004   \n",
      "35552    7.124561    7.290055    7.296061    7.149179    7.215749    7.196333   \n",
      "35553    3.134968    3.256094    3.179739    3.188076    3.144330    3.207743   \n",
      "35554    3.291738    3.392518    3.354275    3.409771    3.488536    3.471732   \n",
      "35555    4.899230    4.727363    4.816136    4.817920    4.805248    4.840851   \n",
      "\n",
      "       GSM3101227  GSM3101228  GSM3101229  \n",
      "35551    7.916520    8.013821    7.798217  \n",
      "35552    7.275178    7.244657    6.918887  \n",
      "35553    3.153569    3.158315    3.186455  \n",
      "35554    3.400055    3.332553    3.312330  \n",
      "35555    4.777877    4.794845    4.739567  \n",
      "\n",
      "[5 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "ds_4 = ds_4.drop(ds_4.index[-1])\n",
    "\n",
    "print(ds_4.tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "c4caa2c9-097d-4198-95cf-8da2b2ef2a0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values in each column:\n",
      " ID_REF        0\n",
      "GSM3101204    0\n",
      "GSM3101205    0\n",
      "GSM3101206    0\n",
      "GSM3101207    0\n",
      "GSM3101208    0\n",
      "GSM3101209    0\n",
      "GSM3101210    0\n",
      "GSM3101211    0\n",
      "GSM3101212    0\n",
      "GSM3101213    0\n",
      "GSM3101214    0\n",
      "GSM3101215    0\n",
      "GSM3101216    0\n",
      "GSM3101217    0\n",
      "GSM3101218    0\n",
      "GSM3101219    0\n",
      "GSM3101220    0\n",
      "GSM3101221    0\n",
      "GSM3101222    0\n",
      "GSM3101223    0\n",
      "GSM3101224    0\n",
      "GSM3101225    0\n",
      "GSM3101226    0\n",
      "GSM3101227    0\n",
      "GSM3101228    0\n",
      "GSM3101229    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check for missing values\n",
    "missing_values_4 = ds_4.isnull().sum()\n",
    "print(\"Missing Values in each column:\\n\", missing_values_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "dbc189b4-d5c4-4cd3-b271-68eab13b6783",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Duplicate columns: []\n"
     ]
    }
   ],
   "source": [
    "duplicate_columns = df.T.duplicated()\n",
    "\n",
    "# Print the duplicate columns\n",
    "duplicate_columns_list = df.columns[duplicate_columns].tolist()\n",
    "print(\"Duplicate columns:\", duplicate_columns_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e55d6fb-fa1c-424c-8645-bdbbd22976ae",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.8.1 Dataset 4 Key Concepts</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "**RMA Signal:** A Robust Multi-array Average is a processed value that represents the intensity of RNA expression on a log2 scale in this study. This is derived from the Affymetrix GeneChip Mouse Gene 1.1 ST peg arrays used in the study to analyze the expression of various genes in the colonic mucosa and submucosa of the mice test groups<a href=\"#mla_159\"> [159]</a>.\n",
    "\n",
    "**RNA Transcript:** An RNA transcript is the RNA strand that is produced when a gene is transcribed. Precursor mRNA (pre-RNA) is one type of RNA transcript. Pre-mRNA is processed into mature mRNA which in turn is translated into a protein.<a href=\"#mla_160\"> [160]</a>.\n",
    "\n",
    "**Affymetrix GeneChip Mouse:** A transcript array to measure RNA transcripts from gene expressions that were used in the study of this dataset<a href=\"#mla_161\"> [161]</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26c7bed7-1ba7-4e95-85db-d4c76e8a4021",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "    <br>\n",
    "    <h4>1.3.8.2 Dataset 4 Abstract</h4>\n",
    "    <hr style=\"width: 75%;\"> <!-- The line is also centered, and you can adjust the width as needed -->\n",
    "    <br>\n",
    "</div>\n",
    "\n",
    "\n",
    "Worsening colonic health is associated with age-related pathologies. In this study, it aimed to \"reveal potential interactions between determinants of colonic health in aging C57BL/6J mice. Analysis of gut microbiota composition revealed an enrichment of various potential pathobionts, including Desulfovibrio spp., and a decline of the health-promoting Akkermansia spp. and Lactobacillus spp. during aging\"<a href=\"#mla_156\"> [156]</a>. The study also found increased gut permeability at higher ages as signs of increased inflammation and its related pathologies.\n",
    "\n",
    "\n",
    "\"This study demonstrates that aging is associated with pronounced changes in gut microbiota composition and colonic gene expression. Furthermore, the strong correlations between specific bacterial genera and host gene expression may imply that orchestrated interactions take place in the vicinity of the colonic wall and potentially mediate colonic health during aging\"<a href=\"#mla_157\"> [157]</a>. There where different gene expressions that occurred during different stages of the test specimen life, which where related to the microbiota composition at the time of life of a specimen.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "<h4>Dataset 4 (columns):</h4>  \n",
    "\n",
    "- 1.) ID_REF: intensity of RNA expression on a log2 scale.\n",
    "- 2.) GSM_XYZ: A mouse test group of many mice and their associated RNA expressions under them as values.\n",
    "\n",
    "<h4>Conclusion</h4>\n",
    "\n",
    "\n",
    "There are 25 mouse test groups that were split into different age spans. It was hypothesized that at older age spans, higher intestinal permeability which is linked to opportunistic microbiotia, affects colonic gene expression. Thousands of samples of colonic gene expression were measured as intra-luminal samples in RNA tests, which is the data under the mice groups<a href=\"#mla_158\"> [158]</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "017c35e0-8214-4d12-a162-69295cfbe741",
   "metadata": {},
   "source": [
    "MLA Works Cited:\n",
    "<ol>\n",
    "<li><a id=\"mla_1\">[1]</a>, <a id=\"mla_2\">[2]</a>, <a id=\"mla_3\">[3]</a>, <a id=\"mla_4\">[4]</a>, <a id=\"mla_5\">[5]</a>: Alzheimer's Association. \"Public Health Approach.\" Alzheimer's Association, https://www.alz.org/professionals/public-health/public-health-approach. Accessed 18 July 2024.</li>\n",
    "<li><a id=\"mla_6\">[6]</a>: Alzheimer's Association. 2024 Alzheimer’s Disease Facts and Figures. Alzheimer’s Association, 2024, https://www.alz.org/media/Documents/alzheimers-facts-and-figures.pdf. Accessed 14 Aug. 2024.</li>\n",
    "<li><a id=\"mla_7\">[7]</a>: Scholistico. \"Naturopathy vs Conventional Medicine: A Comparison.\" Scholistico, https://scholistico.com/naturopathy-vs-conventional-medicine-comparison/. Accessed 14 Aug. 2024.</li> \n",
    "<li><a id=\"mla_8\">[8]</a>: Snider, Pamela, and Jared Zeff. “Unifying Principles of Naturopathic Medicine Origins and Definitions.” Integrative Medicine (Encinitas, Calif.), U.S. National Library of Medicine, Aug. 2019, www.ncbi.nlm.nih.gov/pmc/articles/PMC7219457/.</li>\n",
    "<li><a id=\"mla_9\">[9]</a>: Xie, Junhua, et al. “The Impact of Systemic Inflammation on Alzheimer’s Disease Pathology.” Frontiers in Immunology, vol. 12, 6 Jan. 2022, https://doi.org/10.3389/fimmu.2021.796867.</li>\n",
    "<li><a id=\"mla_10\">[10]</a><a id=\"mla_29\">, [29]</a>: Brisbane Livewell Clinic. “Brisbane Livewell Clinic.” Brisbane Livewell Clinic, 6 Dec. 2020, www.brisbanelivewellclinic.com.au/how-supplements-work-in-naturopathy/#:~:text=A%20Naturopath%20will%20tailor%20a. Accessed 15 Aug. 2024. </li>\n",
    "<li><a id=\"mla_11\">[11]</a>: Winter, Sebastian, and Andreas J Bäumler. “Gut Dysbiosis: Ecological Causes and Causative Effects on Human Disease.” Proceedings of the National Academy of Sciences of the United States of America, vol. 120, no. 50, 4 Dec. 2023, https://doi.org/10.1073/pnas.2316579120. Accessed 11 Jan. 2024.</li>\n",
    "<li><a id=\"mla_12\">[12]</a>: Tongthong Tongthong, et al. “Lacticaseibacillus Rhamnosus Dfa1 Attenuate Cecal Ligation-Induced Systemic Inflammation through the Interference in Gut Dysbiosis, Leaky Gut, and Enterocytic Cell Energy.” International Journal of Molecular Sciences, vol. 24, no. 4, 13 Feb. 2023, pp. 3756–3756, www.ncbi.nlm.nih.gov/pmc/articles/PMC9960508/, https://doi.org/10.3390/ijms24043756. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_13\">[13]</a>: Panpetch, Wimonrat, et al. “Additional Candida Albicans Administration Enhances the Severity of Dextran Sulfate Solution Induced Colitis Mouse Model through Leaky Gut-Enhanced Systemic Inflammation and Gut-Dysbiosis but Attenuated by Lactobacillus Rhamnosus L34.” Gut Microbes, vol. 11, no. 3, pp. 465–480, www.ncbi.nlm.nih.gov/pmc/articles/PMC7527076/, https://doi.org/10.1080/19490976.2019.1662712.</li>\n",
    "<li><a id=\"mla_14\">[14]</a>: Alengebawy, Ahmed, et al. “Heavy Metals and Pesticides Toxicity in Agricultural Soil and Plants: Ecological Risks and Human Health Implications.” Toxics, vol. 9, no. 3, 25 Feb. 2021, p. 42, https://doi.org/10.3390/toxics9030042.</li>\n",
    "<li><a id=\"mla_15\">[15]</a>: Zhang, Huiling, et al. “Association between Heavy Metals Exposure and Persistent Infections: The Mediating Role of Immune Function.” Frontiers in Public Health, vol. 12, 22 July 2024, www.ncbi.nlm.nih.gov/pmc/articles/PMC11298456/, https://doi.org/10.3389/fpubh.2024.1367644. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_16\">[16]</a>: Hope, Janette. “A Review of the Mechanism of Injury and Treatment Approaches for Illness Resulting from Exposure to Water-Damaged Buildings, Mold, and Mycotoxins.” The Scientific World Journal, vol. 2013, 2013, pp. 1–20, www.hindawi.com/journals/tswj/2013/767482/, https://doi.org/10.1155/2013/767482.</li>\n",
    "<li><a id=\"mla_17\">[17]</a>: Krolick, Kristen N., et al. “Effects of Estrogens on Central Nervous System Neurotransmission: Implications for Sex Differences in Mental Disorders.” Progress in Molecular Biology and Translational Science, vol. 160, 2018, pp. 105–171, www.ncbi.nlm.nih.gov/pmc/articles/PMC6737530/, https://doi.org/10.1016/bs.pmbts.2018.07.008.</li>\n",
    "<li><a id=\"mla_18\">[18]</a>: Villa, Alessandro, et al. “Estrogens, Neuroinflammation, and Neurodegeneration.” Endocrine Reviews, vol. 37, no. 4, 1 Aug. 2016, pp. 372–402, www.ncbi.nlm.nih.gov/pmc/articles/PMC4971309/, https://doi.org/10.1210/er.2016-1007.\n",
    "‌</li>\n",
    "<li><a id=\"mla_19\">[19]</a>: Nappi, Rossella E., et al. “Role of Estrogens in Menstrual Migraine.” Cells, vol. 11, no. 8, 15 Apr. 2022, p. 1355, https://doi.org/10.3390/cells11081355.</li>\n",
    "<li><a id=\"mla_20\">[20]</a>: Lies Langouche, et al. “The Hypothalamus-Pituitary-Adrenocortical Response to Critical Illness: A Concept in Need of Revision.” Endocrine Reviews, vol. 44, no. 6, 6 July 2023, pp. 1096–1106, www.ncbi.nlm.nih.gov/pmc/articles/PMC10638597/, https://doi.org/10.1210/endrev/bnad021. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_21\">[21]</a>: Patani, Anil, et al. “Harnessing the Power of Nutritional Antioxidants against Adrenal Hormone Imbalance-Associated Oxidative Stress.” Frontiers in Endocrinology, vol. 14, 30 Nov. 2023, p. 1271521, www.ncbi.nlm.nih.gov/pmc/articles/PMC10720671/#:~:text=As%20an%20antioxidant%2C%20Vitamin%20E, https://doi.org/10.3389/fendo.2023.1271521</li>\n",
    "<li><a id=\"mla_22\">[22]</a>: Zhang, Kai, et al. “Single-Cell Atlas of Murine Adrenal Glands Reveals Immune-Adrenal Crosstalk during Systemic Candida Albicans Infection.” Frontiers in Immunology, vol. 13, 1 Nov. 2022, https://doi.org/10.3389/fimmu.2022.966814. Accessed 9 Jan. 2023.</li>\n",
    "<li><a id=\"mla_23\">[23]</a>: Kandimalla, Ramesh, and P Hemachandra Reddy. “Therapeutics of Neurotransmitters in Alzheimer’s Disease.” Journal of Alzheimer’s Disease : JAD, vol. 57, no. 4, 2017, pp. 1049–1069, www.ncbi.nlm.nih.gov/pubmed/28211810, https://doi.org/10.3233/JAD-161118.</li>\n",
    "<li><a id=\"mla_24\">[24]</a>: Chen, Yijing, et al. “Regulation of Neurotransmitters by the Gut Microbiota and Effects on Cognition in Neurological Disorders.” Nutrients, vol. 13, no. 6, 19 June 2021, p. 2099, https://doi.org/10.3390/nu13062099.</li>\n",
    "<li><a id=\"mla_25\">[25]</a>: Hodo, Thomas Wesley, et al. “Critical Neurotransmitters in the Neuroimmune Network.” Frontiers in Immunology, vol. 11, 21 Aug. 2020, https://doi.org/10.3389/fimmu.2020.01869.</li>\n",
    "<li><a id=\"mla_26\">[26]</a>: Maria Victoria Fernandez, et al. “Genetic and Multi-Omic Resources for Alzheimer Disease and Related Dementia from the Knight Alzheimer Disease Research Center.” Scientific Data, vol. 11, no. 1, 12 July 2024, www.ncbi.nlm.nih.gov/pmc/articles/PMC11245521/, https://doi.org/10.1038/s41597-024-03485-9. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_27\">[27]</a>: Grzadkowski, Michal R., et al. “Systematic Interrogation of Mutation Groupings Reveals Divergent Downstream Expression Programs within Key Cancer Genes.” BMC Bioinformatics, vol. 22, no. 1, 6 May 2021, www.ncbi.nlm.nih.gov/pmc/articles/PMC8101181/, https://doi.org/10.1186/s12859-021-04147-y.</li>\n",
    "<li><a id=\"mla_28\">[28]</a>: Zhao, Enyang, et al. “Comprehensive Characterization of Immune- and Inflammation-Associated Biomarkers Based on Multi-Omics Integration in Kidney Renal Clear Cell Carcinoma.” Journal of Translational Medicine, vol. 17, no. 1, 27 May 2019, www.ncbi.nlm.nih.gov/pmc/articles/PMC6537414/, https://doi.org/10.1186/s12967-019-1927-y. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_30\">[30]</a>: “Naturopathic Approach to Testosterone Replacement Therapy.” NatureMed Clinic, 28 Apr. 2021, www.naturemedclinic.com/a-naturopathic-approach-to-testosterone-replacement-therapy/. Accessed 15 Aug. 2024.</li>\n",
    "<li><a id=\"mla_31\">[31]</a>: Ali, Javed, et al. “Exploring Scientifically Proven Herbal Aphrodisiacs.” Pharmacognosy Reviews, vol. 7, no. 1, 2013, p. 1, https://doi.org/10.4103/0973-7847.112832.</li>\n",
    "<li><a id=\"mla_32\">[32]</a>: AANMC. “Naturopathic Approaches to Gut Health.” AANMC, 7 Nov. 2019, aanmc.org/naturopathic-news/naturopathic-approaches-to-gut-health/</li>\n",
    "<li><a id=\"mla_33\">[33]</a>: Johns Hopkins Medicine. “Chinese Medicine.” Johns Hopkins Medicine, 2019, www.hopkinsmedicine.org/health/wellness-and-prevention/chinese-medicine.</li>\n",
    "<li><a id=\"mla_34\">[34]</a><a id=\"mla_35\">, [35]</a><a id=\"mla_36\">, [36]</a>: Mount Sinai. “Traditional Chinese Medicine Information | Mount Sinai - New York.” Mount Sinai Health System, www.mountsinai.org/health-library/treatment/traditional-chinese-medicine.</li>\n",
    "<li><a id=\"mla_37\">[37]</a>: Wikipedia Contributors. “Naturopathy.” Wikipedia, Wikimedia Foundation, 22 Mar. 2024, en.wikipedia.org/wiki/Naturopathy#References. Accessed 16 Aug. 2024.</li>\n",
    "<li><a id=\"mla_38\">[38]</a><a id=\"mla_39\">, [39]</a>: Wikipedia Contributors. “Traditional Chinese Medicine.” Wikipedia, Wikimedia Foundation, 4 Apr. 2019, en.wikipedia.org/wiki/Traditional_Chinese_medicine.</li>\n",
    "<li><a id=\"mla_40\">[40]</a><a id=\"mla_41\">, [41]</a><a id=\"mla_42\">, [42]</a><a id=\"mla_43\">, [43]</a>: Harvard Library. “Humoral Theory.” Contagion - CURIOSity Digital Collections, Harvard Library, curiosity.lib.harvard.edu/contagion/feature/humoral-theory.</li>\n",
    "<li><a id=\"mla_44\">[44]</a><a id=\"mla_59\">, [59]</a><a id=\"mla_70\">, [70]</a><a id=\"mla_88\">, [88]</a><a id=\"mla_108\">, [108]</a>: Wikipedia Contributors. “Humorism.” Wikipedia, Wikimedia Foundation, 22 May 2019, en.wikipedia.org/wiki/Humorism.</li>\n",
    "<li><a id=\"mla_45\">[45]</a><a id=\"mla_46\">, [46]</a><a id=\"mla_52\">, [52]</a>: Wikipedia Contributors. “Parasympathetic Nervous System.” Wikipedia, Wikimedia Foundation, 4 Jan. 2020, en.wikipedia.org/wiki/Parasympathetic_nervous_system.</li>\n",
    "<li><a id=\"mla_47\">[47]</a><a id=\"mla_48\">, [48]</a><a id=\"mla_49\">, [49]</a><a id=\"mla_50\">, [50]</a><a id=\"mla_51\">, [51]</a>: Wikipedia Contributors. “Sympathetic Nervous System.” Wikipedia, Wikimedia Foundation, 11 Aug. 2019, en.wikipedia.org/wiki/Sympathetic_nervous_system.</li>\n",
    "<li><a id=\"mla_53\">[53]</a><a id=\"mla_67\">, [67]</a>: “The Four Stages of Adrenal Fatigue – Total Health Chiropractic Cleveland Blog.” A-Familychiropractic.com, 30 Nov. 2019, www.a-familychiropractic.com/blog/the-four-stages-of-adrenal-fatigue/#:~:text=Stage%203%20(Meet%20the%20Resistance)&text=Almost%20simultaneously%20once%20in%20this. Accessed 16 Aug. 2024.</li>\n",
    "<li><a id=\"mla_54\">[54]</a><a id=\"mla_55\">, [55]</a><a id=\"mla_56\">, [56]</a><a id=\"mla_57\">, [57]</a><a id=\"mla_58\">, [58]</a><a id=\"mla_62\">, [62]</a><a id=\"mla_64\">, [64]</a><a id=\"mla_65\">, [65]</a><a id=\"mla_66\">, [66]</a><a id=\"mla_68\">, [68]</a><a id=\"mla_69\">, [69]</a><a id=\"mla_78\">, [78]</a><a id=\"mla_79\">, [79]</a><a id=\"mla_80\">, [80]</a><a id=\"mla_81\">, [81]</a><a id=\"mla_82\">, [82]</a><a id=\"mla_83\">, [83]</a><a id=\"mla_89\">, [89]</a><a id=\"mla_90\">, [90]</a><a id=\"mla_91\">, [91]</a><a id=\"mla_92\">, [92]</a><a id=\"mla_93\">, [93]</a><a id=\"mla_94\">, [94]</a><a id=\"mla_95\">, [95]</a><a id=\"mla_96\">, [96]</a>: Ripperger, Chad. Introduction to the Science of Mental Health. 1 May 2001.</li>\n",
    "<li><a id=\"mla_60\">[60]</a>: Wikipedia Contributors. “Gallbladder.” Wikipedia, Wikimedia Foundation, 2 July 2019, en.wikipedia.org/wiki/Gallbladder.</li>\n",
    "<li><a id=\"mla_61\">[61]</a>: Calhoun, Ari. “Kith & Kin Wellness.” Kith & Kin Wellness, 13 May 2020, www.kithandkinwellness.com/blog-posts/phase-two-detoxification#:~:text=Phase%202%20detox%20is%20known. Accessed 16 Aug. 2024.</li>\n",
    "<li><a id=\"mla_63\">[63]</a>: “Everyday Toxicology – Absorbing, Processing & Eliminating Ingredients.” Center for Research on Ingredient Safety, www.canr.msu.edu/news/everyday-toxicology-absorbing-processing-eliminating-ingredients.</li>\n",
    "<li><a id=\"mla_71\">[71]</a>: Kenli, Kenli. Black Bile Humor | Definition, Disorders & Treatment. 21 Nov. 2023, study.com/academy/lesson/black-bile-humor-definition-lesson-quiz.html#:~:text=In%20Humorism%2C%20Black%20Bile%20was,harmful%20effects%20may%20take%20place. Accessed 16 Aug. 2024.</li>\n",
    "<li><a id=\"mla_72\">[72]</a><a id=\"mla_73\">, [73]</a><a id=\"mla_74\">, [74]</a><a id=\"mla_75\">, [75]</a><a id=\"mla_76\">, [76]</a><a id=\"mla_77\">, [77]</a>: Cleveland Clinic. “Spleen: Spleen Function, Enlarged Spleen, What Does the Spleen Do.” Cleveland Clinic, 29 Apr. 2021, my.clevelandclinic.org/health/body/21567-spleen.</li>\n",
    "<li><a id=\"mla_84\">[84]</a>: Fischer, Adrian G., and Markus Ullsperger. “An Update on the Role of Serotonin and Its Interplay with Dopamine for Reward.” Frontiers in Human Neuroscience, vol. 11, 11 Oct. 2017, https://doi.org/10.3389/fnhum.2017.00484.</li>\n",
    "<li><a id=\"mla_85\">[85]</a>: Bromberg-Martin, Ethan S., et al. “Dopamine in Motivational Control: Rewarding, Aversive, and Alerting.” Neuron, vol. 68, no. 5, 9 Dec. 2010, pp. 815–834, www.ncbi.nlm.nih.gov/pmc/articles/PMC3032992/, https://doi.org/10.1016/j.neuron.2010.11.022.</li>\n",
    "<li><a id=\"mla_86\">[86]</a>: Dfarhud, Dariush, et al. “Happiness & Health: The Biological Factors- Systematic Review Article.” Iranian Journal of Public Health, vol. 43, no. 11, Nov. 2014, pp. 1468–77, www.ncbi.nlm.nih.gov/pmc/articles/PMC4449495/‌</li>\n",
    "<li><a id=\"mla_87\">[87]</a>: Casado-Sainz, Agata, et al. “Dorsal Striatal Dopamine Induces Fronto-Cortical Hypoactivity and Attenuates Anxiety and Compulsive Behaviors in Rats.” Neuropsychopharmacology, vol. 47, no. 2, 1 Jan. 2022, pp. 454–464, www.nature.com/articles/s41386-021-01207-y, https://doi.org/10.1038/s41386-021-01207-y</li>\n",
    "<li><a id=\"mla_97\">[97]</a>: Anthenelli, Robert M., and Marc A. Schuckit. “Genetic Studies of Alcoholism.” International Journal of the Addictions, vol. 25, no. sup1, Jan. 1990, pp. 81–94, https://doi.org/10.3109/10826089009067006. Accessed 26 Mar. 2021.</li>\n",
    "<li><a id=\"mla_98\">[98]</a>: Panchenko, Polina E., et al. “[Epigenetics and Nutrition: Maternal Nutrition Impacts on Placental Development and Health of Offspring].” Biologie Aujourd’hui, vol. 209, no. 2, 2015, pp. 175–187, pubmed.ncbi.nlm.nih.gov/26514387/, https://doi.org/10.1051/jbio/2015021</li>\n",
    "<li><a id=\"mla_99\">[99]</a><a id=\"mla_103\">, [103]</a>: Lowen, Alexander. Bioenergetics. New York, Penguin/Arkana, 1994.</li>\n",
    "<li><a id=\"mla_100\">[100]</a><a id=\"mla_102\">, [102]</a>: Wikipedia Contributors. “Wilhelm Reich.” Wikipedia, Wikimedia Foundation, 9 Nov. 2019, en.wikipedia.org/wiki/Wilhelm_Reich</li>\n",
    "<li><a id=\"mla_101\">[101]</a>: Wikipedia Contributors. “Sigmund Freud.” Wikipedia, Wikimedia Foundation, 17 Feb. 2019, en.wikipedia.org/wiki/Sigmund_Freud</li>\n",
    "<li><a id=\"mla_104\">[104]</a>: Whellan, David J., and Christopher M. O’Connor. “Function Follows Form.” JACC: Heart Failure, vol. 9, no. 7, July 2021, pp. 482–483, https://doi.org/10.1016/j.jchf.2021.02.012. Accessed 22 Aug. 2022.</li>\n",
    "<li><a id=\"mla_105\">[105]</a>: Our Guiding Principles — Matrix Wellness. “Matrix Wellness.” Matrix Wellness, Sept. 2022, www.matrixwellness.com/oracle/the-therapeutic-order. Accessed 17 Aug. 2024.</li>\n",
    "<li><a id=\"mla_106\">[106]</a>: “The Therapeutic Order of Naturopathic Medicine.” Balance into Health, www.balanceintohealth.com/blog/the-therapeutic-order-of-naturopathic-medicine.</li>\n",
    "<li><a id=\"mla_107\">[107]</a>: “Yin and Yang as Journey toward Harmony.” The Hero’s Journey: Life’s Great Adventure, 18 Apr. 2020, www.yourheroicjourney.com/yin-and-yang-as-journey-toward-harmony-2/.</li>\n",
    "<li><a id=\"mla_109\">[109]</a><a id=\"mla_110\">, [110]</a><a id=\"mla_111\">, [111]</a><a id=\"mla_112\">, [112]</a><a id=\"mla_113\">, [113]</a><a id=\"mla_114\">, [114]</a>: Wikipedia Contributors. “Four Temperaments.” Wikipedia, Wikimedia Foundation, 26 Feb. 2019, en.wikipedia.org/wiki/Four_temperaments.</li>\n",
    "<li><a id=\"mla_115\">[115]</a><a id=\"mla_117\">, [117]</a>: Kinney, Jefferson W., et al. “Inflammation as a Central Mechanism in Alzheimer’s Disease.” Alzheimer’s & Dementia: Translational Research & Clinical Interventions, vol. 4, no. 1, 2018, pp. 575–590, www.ncbi.nlm.nih.gov/pmc/articles/PMC6214864/, https://doi.org/10.1016/j.trci.2018.06.014.</li>\n",
    "<li><a id=\"mla_116\">[116]</a><a id=\"mla_118\">, [118]</a>: Chen, Linlin, et al. “Inflammatory Responses and Inflammation-Associated Diseases in Organs.” Oncotarget, vol. 9, no. 6, 14 Dec. 2018, pp. 7204–7218, https://doi.org/10.18632/oncotarget.23208.</li>\n",
    "<li><a id=\"mla_119\">[119]</a><a id=\"mla_120\">, [120]</a>: HUANG, WEN-JUAN, et al. “Role of Oxidative Stress in Alzheimer’s Disease.” Biomedical Reports, vol. 4, no. 5, 15 Mar. 2016, pp. 519–522, www.ncbi.nlm.nih.gov/pmc/articles/PMC4840676/, https://doi.org/10.3892/br.2016.630.</li>\n",
    "<li><a id=\"mla_121\">[121]</a><a id=\"mla_122\">, [122]</a>: Al Bander, Zahraa, et al. “The Gut Microbiota and Inflammation: An Overview.” International Journal of Environmental Research and Public Health, vol. 17, no. 20, 1 Oct. 2020, www.ncbi.nlm.nih.gov/pmc/articles/PMC7589951/, https://doi.org/10.3390/ijerph17207618.</li>\n",
    "<li><a id=\"mla_123\">[123]</a><a id=\"mla_124\">, [124]</a>: Kunst, Claudia, et al. “The Influence of Gut Microbiota on Oxidative Stress and the Immune System.” Biomedicines, vol. 11, no. 5, 8 May 2023, p. 1388, https://doi.org/10.3390/biomedicines11051388.</li>\n",
    "<li><a id=\"mla_125\">[125]</a><a id=\"mla_126\">, [126]</a><a id=\"mla_127\">, [127]</a><a id=\"mla_128\">, [128]</a>: Fu, Xin, et al. “Danggui-Shaoyao-San: New Hope for Alzheimer’s Disease.” Aging and Disease, vol. 7, no. 4, 2016, p. 502, www.ncbi.nlm.nih.gov/pmc/articles/PMC4963193/, https://doi.org/10.14336/ad.2015.1220. Accessed 9 Feb. 2023.</li>\n",
    "<li><a id=\"mla_129\">[129]</a>: Liu, Piaoxue, et al. “Danggui-Shaoyao-San Attenuates Cognitive Impairment via the Microbiota–Gut–Brain Axis with Regulation of Lipid Metabolism in Scopolamine-Induced Amnesia.” Frontiers in Immunology, vol. 13, 19 May 2022, www.ncbi.nlm.nih.gov/pmc/articles/PMC9162091/, https://doi.org/10.3389/fimmu.2022.796542. Accessed 3 Oct. 2023.</li>\n",
    "<li><a id=\"mla_130\">[130]</a><a id=\"mla_131\">, [131]</a>: He, Jiawei, et al. “Table5_Danggui Shaoyao San: Comprehensive Modulation of the Microbiota-Gut-Brain Axis for Attenuating Alzheimer’s Disease-Related Pathology.XLSX.” Figshare, 12 Jan. 2024, figshare.com/articles/dataset/Table5_Danggui_Shaoyao_San_comprehensive_modulation_of_the_microbiota-gut-brain_axis_for_attenuating_Alzheimer_s_disease-related_pathology_XLSX/24987513?file=44011431, https://doi.org/10.3389/u002Ffphar.2023.1338804.s009. Accessed 19 Aug. 2024.</li>\n",
    "<li><a id=\"mla_132\">[132]</a>: Danggui Shaoyao San: Comprehensive Modulation of the Microbiota-Gut-Brain Axis for Attenuating Alzheimer’s Disease-Related Pathology. 11 Jan. 2024, www.frontiersin.org/journals/pharmacology/articles/10.3389/fphar.2023.1338804/full. Accessed 19 Aug. 2024.</li>\n",
    "<li><a id=\"mla_133\">[133]</a><a id=\"mla_134\">, [134]</a>: How Many Differentially Expressed Genes: A Perspective from the Comparison of Genotypic and Phenotypic Distances. Jan. 2018, www.sciencedirect.com/science/article/pii/S0888754317300745#:~:text=Among%20these%20groups%2C%20the%20expression,molecular%20biomarkers%20of%20different%20phenotypes.</li>\n",
    "<li><a id=\"mla_135\">[135]</a>: Tarca, Adi L., et al. “Analysis of Microarray Experiments of Gene Expression Profiling.” American Journal of Obstetrics and Gynecology, vol. 195, no. 2, Aug. 2006, pp. 373–388, www.ncbi.nlm.nih.gov/pmc/articles/PMC2435252/, https://doi.org/10.1016/j.ajog.2006.07.001.\n",
    "‌</li>\n",
    "<li><a id=\"mla_136\">[136]</a>: “Gene Ontology Overview.” Gene Ontology Resource, geneontology.org/docs/ontology-documentation/.</li>\n",
    "<li><a id=\"mla_137\">[137]</a>: Safari-Alighiarloo, Nahid, et al. “Protein-Protein Interaction Networks (PPI) and Complex Diseases.” Gastroenterology and Hepatology from Bed to Bench, vol. 7, no. 1, 2014, pp. 17–31, www.ncbi.nlm.nih.gov/pmc/articles/PMC4017556/#:~:text=A%20protein%2Dprotein%20interaction%20(PPI. Accessed 20 Aug. 2024.</li>\n",
    "<li><a id=\"mla_138\">[138]</a>: Hub Genes. 2023, www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/hub-genes#:~:text=Hub%20gene%3A%20A%20gene%20that,the%20network%20as%20a%20whole.</li>\n",
    "<li><a id=\"mla_139\">[139]</a>: Bathina, Siresha, and Undurti N. Das. “Brain-Derived Neurotrophic Factor and Its Clinical Implications.” Archives of Medical Science, vol. 11, no. 6, 2015, pp. 1164–1178, https://doi.org/10.5114/aoms.2015.56342.</li>\n",
    "<li><a id=\"mla_140\">[140]</a>: Wikipedia Contributors. “WWTR1.” Wikipedia, Wikimedia Foundation, 20 May 2024, en.wikipedia.org/wiki/WWTR1. Accessed 20 Aug. 2024.</li>\n",
    "<li><a id=\"mla_141\">[141]</a>: “Braak Staging.” Wikipedia, 13 Dec. 2020, en.wikipedia.org/wiki/Braak_staging.</li>\n",
    "<li><a id=\"mla_142\">[142]</a>: Wikipedia Contributors. “Amyloid Beta.” Wikipedia, Wikimedia Foundation, 24 Mar. 2019, en.wikipedia.org/wiki/Amyloid_beta.</li>\n",
    "<li><a id=\"mla_143\">[143]</a>: Wikipedia Contributors. “Beta-Secretase 1.” Wikipedia, Wikimedia Foundation, 19 Oct. 2018, en.wikipedia.org/wiki/Beta-secretase_1.</li>\n",
    "<li><a id=\"mla_144\">[144]</a><a id=\"mla_145\">, [145]</a><a id=\"mla_146\">, [146]</a><a id=\"mla_147\">, [147]</a>: Yu, Wuhan, et al. “Table_4_Exploring the Key Genes and Identification of Potential Diagnosis Biomarkers in Alzheimer’s Disease Using Bioinformatics Analysis.XLSX.” Figshare, 14 June 2021, figshare.com/articles/dataset/Table_4_Exploring_the_Key_Genes_and_Identification_of_Potential_Diagnosis_Biomarkers_in_Alzheimer_s_Disease_Using_Bioinformatics_Analysis_XLSX/14777766/1?file=28396890, https://doi.org/10.3389/u002Ffnagi.2021.602781.s009. Accessed 20 Aug. 2024</li>\n",
    "<li><a id=\"mla_148\">[148]</a><a id=\"mla_149\">, [149]</a><a id=\"mla_150\">, [150]</a>: “The Impact of Rhodiola Rosea on Biomarkers of Diabetes, Inflammation, and Microbiota in a Leptin Receptor-Knockout Mouse Model.” Https://Datadryad.org/Stash/Dataset/Doi:10.7280/D1FX2D, 16 June 2022.</li>\n",
    "<li><a id=\"mla_151\">[151]</a><a id=\"mla_152\">, [152]</a>: “Rhodiola: Uses, Side Effects, Interactions, Dosage, and Warning.” Webmd.com, 2009, www.webmd.com/vitamins/ai/ingredientmono-883/rhodiola.</li>\n",
    "<li><a id=\"mla_153\">[153]</a>: Wikipedia Contributors. “C-Reactive Protein.” Wikipedia, Wikimedia Foundation, 15 Feb. 2019, en.wikipedia.org/wiki/C-reactive_protein.</li>\n",
    "<li><a id=\"mla_154\">[154]</a>: “Glyceraldehyde 3-Phosphate Dehydrogenase.” Wikipedia, 23 Sept. 2019, en.wikipedia.org/wiki/Glyceraldehyde_3-phosphate_dehydrogenase.</li>\n",
    "<li><a id=\"mla_155\">[155]</a>: “Cycle Threshold - an Overview | ScienceDirect Topics.” Www.sciencedirect.com, www.sciencedirect.com/topics/immunology-and-microbiology/cycle-threshold.</li>\n",
    "<li><a id=\"mla_156\">[156]</a><a id=\"mla_157\">, [157]</a><a id=\"mla_158\">, [158]</a>: “GSE113257 - Integrative Analysis of Gut Microbiota Composition, Host Colonic Gene Expression and Intraluminal Metabolites in Aging C57BL/6J Mice - OmicsDI.” Omicsdi.org, 2018, www.omicsdi.org/dataset/geo/GSE113257. Accessed 20 Aug. 2024.</li>\n",
    "<li><a id=\"mla_159\">[159]</a>: Dnastar.com, 2022, www.dnastar.com/arraystar_help/index.html#.</li>\n",
    "<li><a id=\"mla_160\">[160]</a>: “RNA Transcript.” Toolkit, toolkit.ncats.nih.gov/glossary/rna-transcript/#:~:text=An%20RNA%20transcript%20is%20the.</li>\n",
    "<li><a id=\"mla_161\">[161]</a>: “GeneChipTM Mouse Gene 1.1 ST Array Plate.” Thermofisher.com, 2024, www.thermofisher.com/order/catalog/product/901418. Accessed 20 Aug. 2024.</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cf6277b-e150-431b-b2ee-9cad808fd32e",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
